\documentclass[12pt]{article}

\usepackage[english]{babel}
\usepackage{amsmath,amssymb,bm}

\usepackage[table]{xcolor}

\usepackage{authblk}

\usepackage{natbib}
\bibliographystyle{apa}	
\usepackage[colorlinks=true, linkcolor=blue, citecolor=blue, urlcolor=blue]{hyperref}

\usepackage{graphicx}
\graphicspath{{plots/}}
\usepackage{grffile}

\usepackage{tikz}
\usetikzlibrary{positioning,calc,intersections,arrows,decorations.pathreplacing}
\usepackage{pgfplots}
\pgfplotsset{compat=1.12}

\usepackage{ccaption}

\definecolor{tab10_blue}{HTML}{1F77B4}
\definecolor{tab10_orange}{HTML}{FF7F0E}
\definecolor{tab10_green}{HTML}{2CA02C}

\title{Estimating dispersal rates and locating genetic ancestors with genome-wide genealogies}
\author[1]{Matthew M Osmond}
\author[2]{Graham Coop}
\affil[1]{Department of Ecology \& Evolutionary Biology, University of Toronto}
\affil[2]{Department of Evolution \& Ecology and Center for Population Biology, University of California - Davis}
\date{}

\thispagestyle{plain}
\begin{document}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\maketitle
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Abstract}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Spatial patterns in genetic diversity are shaped by individuals dispersing from their parents and larger-scale population movements. It has long been appreciated that these patterns of movement shape the underlying genealogies along the genome leading to geographic patterns of isolation by distance in contemporary population genetic data. However, extracting the enormous amount of information contained in genealogies along recombining sequences has, until recently, not been computationally feasible. Here we capitalize on important recent advances in genome-wide gene-genealogy reconstruction and develop methods to use thousands of trees to estimate per-generation dispersal rates and to locate the genetic ancestors of a sample back through time. We take a likelihood approach in continuous space using a simple approximate model (branching Brownian motion) as our prior distribution of spatial genealogies. After testing our method with simulations we apply it to \textit{Arabidopsis thaliana}. We estimate a dispersal rate of roughly $60$km$^2$ per generation, slightly higher across latitude than across longitude, potentially reflecting a northward post-glacial expansion. Locating ancestors allows us to visualize major geographic movements, alternative geographic histories, and admixture. Our method highlights the huge amount of information about past dispersal events and population movements contained in genome-wide genealogies.  

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Introduction}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Patterns of genetic diversity are shaped by the movements of individuals, as individuals move their alleles around the landscape as they disperse. Patterns of individual movement reflect individual-level dispersal; children move away from their parents' village and dandelion seeds blow in the wind. These patterns also reflect large-scale movements of populations. For example, in the past decade we have learnt about the large-scale movement of different peoples across the world from ancient DNA \citep{slatkin2016ancient,reich2018we}. Such large scale movements of individuals also occur in other species during biological invasions or with the retreat and expansion of populations in and out of glacial refugia, tracking the waxing and waning of the ice ages \citep{hewitt2000genetic}.

An individual's set of genealogical ancestors expands rapidly geographically back through time in sexually reproducing organisms \citep{KELLEHER20161,Coop_where_genetic}. Due to limited recombination each generation, more than a few tens of generations back an individual's genetic ancestors represent only a tiny sample of their vast number of genealogical ancestors \citep{donnelly1983probability,Coop_many_genetic}. Yet the geographic locations of genetic ancestors still represent an incredibly rich source of information on population history \citep{bradburd2019spatial}. We can hope to learn about the geography of genetic ancestors because individuals who are geographically close are often genetically more similar across their genomes; their ancestral lineages have only dispersed for a relatively short time and distance since they last shared a geographically-close common ancestor. This pattern is termed isolation by distance. These ideas about the effects of geography and genealogy have underlain our understanding of spatial population genetics since its inception  \citep{wright1943isolation,malecot1948mathematics}. Under coalescent models, lineages move spatially, as a Brownian motion if dispersal is random and local, splitting to give rise to descendent lineages until we reach the present day. Such models underlie inferences based on increasing allele frequency differentiation (such as $F_{ST}$) with geographic distance \citep{rousset1997genetic} and the drop-off in the sharing of long blocks of genome shared identical by descent among pairs of individuals \citep{ralph2013geography,ringbauer2017inferring}. These models also are the basis of methods that seek violations of isolation-by-distance \citep{wang2014isolation}.

While spatial genealogies have proven incredibly useful for theoretical tools and intuition, with few exceptions they have not proven useful for inferences because we have not been able to construct these genealogies along recombining sequences. In non-recombining chromosomes (e.g., mtDNA and Y), constructed genealogies have successfully been used to understand patterns of dispersal and spatial spread \citep{avise2009phylogeography}. However, these spatial genealogy inferences are necessarily limited as a single genealogy holds only limited information about the history of populations in a recombining species \citep{barton1995genealogies}. Phylogenetic approaches to geography \citep[`phylogeography';][]{knowles2009statistical} have been more widely and successfully applied to pathogens to track the spatial spread of epidemics, such as SARS-Cov-2 \citep{martin2021insights}, but such approaches have yet to be applied to the thousands of genealogical trees that exist in sexual populations. 

Here we capitalize on the recent ability to infer a sequence of genealogies, with branch lengths, across recombining genomes \citep{rasmussen2014genome,speidel2019method,wohns2021unified}. Equipped with this information, we develop a method that uses a sequence of trees to estimate dispersal rates and locate genetic ancestors in continuous, 2-dimensional space under the assumption of Brownian motion. Using thousands of approximately unlinked trees, we multiply likelihoods of the dispersal rate across trees to get a genome-wide estimate and use the sequence of trees to predict a cloud of ancestral locations as a way to visualize geographic ancestries. We first test our approach with simulations and then apply it to \textit{Arabidopsis thaliana}, a species with a wide geographic distribution and a complex history \citep{fulgione2018archaic}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Results}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Overview of approach} 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

We first give an overview of our approach, the major components of which are illustrated in Figure \ref{fig:overview}. See \nameref{sec:methods} for more details.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%overview figure%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]
\begin{center}
\begin{tikzpicture}[>=triangle 60, scale=0.75, every node/.style={scale=0.75}] %if figure too big, adjust both scales equally

%%%%%%%%%%
%tree sequence
%%%%%%%%%%

%%%%%%%%%%%%%%%
%genome
  \node[] (gleft) at (-8,-1) {...};
  \node[] (gright) at (-2.75,-1) {...};
  \draw[thick] (gleft) -- (gright);
  \node[] (gleft) at (-2.5,-1) {};
  \node[] (gright) at (2.5,-1) {};
  \draw[thick] (gleft) -- (gright);
  \node[] (gleft) at (2.75,-1) {...};
  \node[] (gright) at (8,-1) {...};
  \draw[thick] (gleft) -- (gright);
  \coordinate (gleft) at (-8,-1);
  \coordinate (gright) at (8,-1);
  
%%%%%%%%%%%%%%
%focal tree
  
  %location on genome
  \coordinate (tree) at (0,0);
  \draw[line width=2mm, gray] (-0.1,-1) -- (0.1,-1) node[midway, below] (locus) {locus $i$};
  
  % sample nodes
  \foreach \n in {1,...,5}{
    \coordinate (n\n) at (\n - 3, 0); % define coordinate
    \node[below = 0cm of n\n] {\n}; %node label
  }
  % ancestor node 1
  \foreach \t in {0.5}{
    \foreach \n in {4,5}{
      \draw[ultra thick] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ultra thick] (n4 |- 0,\t) -- (n5 |- 0,\t) coordinate[midway] (n6);
  }    
  % ancestor node 2
  \foreach \t in {1}{ %coalescent time
    \foreach \n in {1,2}{ %nodes that coalesce
      \draw[ultra thick] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ultra thick] (n1 |- 0,\t) -- (n2 |- 0,\t) coordinate[midway] (n7);
  }
  % ancestor node 3
  \foreach \t in {1.5}{
    \foreach \n in {3,6}{
      \draw[ultra thick] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ultra thick] (n3 |- 0,\t) -- (n6 |- 0,\t) coordinate[midway] (n8);
  }  
  % ancestor node 4
  \foreach \t in {3}{
    \foreach \n in {7,8}{
      \draw[ultra thick] (n\n) -- ( n\n |- 0,\t );
    }
  \draw[ultra thick] ( n7 |- 0,\t ) -- ( n8 |- 0,\t ) coordinate[midway] (n9);
  }
  % mrca stem
  \draw[ultra thick] (n9) -- (n9 |- 0,3.25);      
 
 %%%%%%%%%%%%%%
 %left tree
  
  %location on genome
  \coordinate (tree) at (0,0);
  \draw[line width=2mm, gray] (-5.6, -1) -- (-5.4, -1) node[midway, below] (llocus) {locus $i-1$};
  
  % sample nodes
  \foreach \n in {1,...,5}{
    \coordinate (n\n) at (-6.5 + \n - 2, 0); % define coordinate
    \node[below = 0cm of n\n, gray] {\n}; %node label
  }
  % ancestor node 1
  \foreach \t in {0.25}{
    \foreach \n in {3,4}{
      \draw[thick, gray] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ thick, gray] (n3 |- 0,\t) -- (n4 |- 0,\t) coordinate[midway] (n6);
  }    
  % ancestor node 2
  \foreach \t in {0.5}{ %coalescent time
    \foreach \n in {6,5}{ %nodes that coalesce
      \draw[ thick, gray] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ thick, gray] (n6 |- 0,\t) -- (n5 |- 0,\t) coordinate[midway] (n7);
  }
  % ancestor node 3
  \foreach \t in {1.75}{
    \foreach \n in {2,7}{
      \draw[ thick, gray] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ thick, gray] (n2 |- 0,\t) -- (n7 |- 0,\t) coordinate[midway] (n8);
  }  
  % ancestor node 4
  \foreach \t in {2.75}{
    \foreach \n in {1,8}{
      \draw[ thick, gray] (n\n) -- ( n\n |- 0,\t );
    }
  \draw[thick, gray] ( n1 |- 0,\t ) -- ( n8 |- 0,\t ) coordinate[midway] (n9);
  }
  % mrca stem
  \draw[thick, gray] (n9) -- (n9 |- 0,3); 
  
%%%%%%%%%%%%%%
 %right tree
  
  %location on genome
  \coordinate (tree) at (0,0);
  \draw[line width=2mm, gray] (5.15,-1) -- (5.35,-1) node[midway, below] (rlocus){locus $i+1$};
%  
  % sample nodes
  \foreach \n in {1,...,5}{
    \coordinate (n\n) at (5.5 + \n - 3, 0); % define coordinate
    \node[below = 0cm of n\n, gray] {\n}; %node label
  }
  % ancestor node 1
  \foreach \t in {0.75}{
    \foreach \n in {1,2}{
      \draw[ thick, gray] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ thick, gray] (n1 |- 0,\t) -- (n2 |- 0,\t) coordinate[midway] (n6);
  }    
  % ancestor node 2
  \foreach \t in {1.25}{ %coalescent time
    \foreach \n in {6,3}{ %nodes that coalesce
      \draw[ thick, gray] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ thick, gray] (n6 |- 0,\t) -- (n3 |- 0,\t) coordinate[midway] (n7);
  }
  % ancestor node 3
  \foreach \t in {2.25}{
    \foreach \n in {7,4}{
      \draw[ thick, gray] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ thick, gray] (n7 |- 0,\t) -- (n4 |- 0,\t) coordinate[midway] (n8);
  }  
  % ancestor node 4
  \foreach \t in {3.25}{
    \foreach \n in {8,5}{
      \draw[ thick, gray] (n\n) -- ( n\n |- 0,\t );
    }
  \draw[thick, gray] ( n8 |- 0,\t ) -- ( n5 |- 0,\t ) coordinate[midway] (n9);
  }
  % mrca stem
  \draw[thick, gray] (n9) -- (n9 |- 0,3.5); 
  
%%%%%%%%%%%%%%%%%
%time cut off
  \draw[thick, dashed] (-8,2) -- (8,2) node[right] {$T$};

%%%%%%%%%%%%
% down arrows
\draw[->] (locus.south) -- ($(locus.south) + (0,-1)$) node[midway, right, yshift=0.1cm] {ignore deep past};
\draw[->, gray] (llocus.south) -- ($(llocus.south) + (0,-1)$) node[below, yshift=-1cm] {...};
\draw[->, gray] (rlocus.south) -- ($(rlocus.south) + (0,-1)$) node[below, yshift=-1cm] {...};

%%%%%%%%%%%%%
%subtrees
%%%%%%%%%%%

  % sample nodes
  \foreach \n in {1,...,5}{
    \coordinate (n\n) at (\n - 3, -4.5); % define coordinate
    \node[below = 0cm of n\n] {\n}; %node label
  }
  % ancestor node 1
  \foreach \t in {-4}{
    \foreach \n in {4,5}{
      \draw[ultra thick, red] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ultra thick, red] (n4 |- 0,\t) -- (n5 |- 0,\t) coordinate[midway] (n6);
  }
  % ancestor node 2
  \foreach \t in {-3.5}{ %coalescent time
%    \foreach \n in {1,2}{ %nodes that coalesce
%      \draw[ultra thick] (n\n) -- (n\n |- 0,\t);
%    }
  \draw[ultra thick, blue] (n1) -- (n1 |- 0,\t)  node[midway, right] {$t_1$}; 
  \draw[ultra thick, blue] (n2) -- (n2 |- 0,\t); 
  \draw[ultra thick, blue] (n1 |- 0,\t) -- (n2 |- 0,\t) coordinate[midway] (n7);
  }
  % ancestor node 3
  \foreach \t in {-3}{
    \foreach \n in {3,6}{
      \draw[ultra thick, red] (n\n) -- (n\n |- 0,\t) node[midway, right] {$t_\n$};
    }
  \draw[ultra thick, red] (n3 |- 0,\t) -- (n6 |- 0,\t) coordinate[midway] (n8);
  }  
  % blue mrca stem
  \draw[ultra thick, blue] (n7) -- (n7 |- 0,-2.75); 
  % red mrca stem
  \draw[ultra thick, red] (n8) -- (n8 |- 0,-2.75); 
  
%  \node[gray] at (1,-6) {subtree $j$};  

% down arrows
\draw[->] ($(n3.south) + (0,-0.75)$) -- ($(n3.south) + (0,-1.75)$) node[midway, right, yshift=0.1cm] {extract shared times};
\draw[->, gray] ($(n3.south) + (-5.5,-0.75)$) -- ($(n3.south) + (-5.5,-1.75)$) node[below, yshift=-1.25cm] {...};
\draw[->, gray] ($(n3.south) + (5.25,-0.75)$) -- ($(n3.south) + (5.25,-1.75)$) node[below, yshift=-1.25cm] {...};

%%%%%%%%%%%%
%shared times
%%%%%%%%%%%

  \node[anchor = north] (matrix) at ($(n3.south) + (0,-2)$) {
   $\begin{pmatrix} 
        \color{blue}{t_1} & \color{blue}{0} & \color{black}{0} & \color{black}{0} & \color{black}{0} \\
        \color{blue}{0} & \color{blue}{t_1} & \color{black}{0}& \color{black}{0} & \color{black}{0} \\
        \color{black}{0} & \color{black}{0} & \color{red}{t_3} & \color{red}{0} & \color{red}{0} \\ 
        \color{black}{0} & \color{black}{0} & \color{red}{0} & \color{red}{t_3} & \color{red}{t_6} \\ 
        \color{black}{0} & \color{black}{0} & \color{red}{0} & \color{red}{t_6} & \color{red}{t_3} 
  \end{pmatrix}$
  };

% down arrows
\draw[decoration={brace,raise=5pt,amplitude=10pt,mirror},decorate] ($(matrix.south) + (gleft) + (0,1.25)$) -- ($(matrix.south) + (gright) + (0,1.25)$);
\coordinate(A) at ($(matrix.south) + (0, -1.5)$);
\draw[->] ($(matrix.south) + (0, -0.5)$) -- (A) node[midway, right, yshift=0.1cm] {estimate dispersal rate};

%%%%%%%%%%%%%%
%dispersal rate
%%%%%%%%%%%%%%

\node[anchor=north] at ($(matrix.south) + (gleft) + (0, -3.5)$) (locations) {
  \begin{axis}[
      title=sample locations,
      title style={at={(axis description cs:0.5,0.95)}, gray},
      cycle list name=black white,
      height=5cm,
      width=5cm,
      xlabel=$x$,
      ylabel=$y$,
      samples=5,
      domain=0:1,
      ymin = 0,
      ymax = 1,
      ticks=none,
      x label style={at={(axis description cs:0.5,0)}, anchor=north, color=gray},
      y label style={at={(axis description cs:0,0.5)},rotate=-90,anchor=east, color=gray}
  ]
%  \addplot+[only marks]  {rand}; %plot some random locations
    \addplot[only marks, mark=text, text mark = {1}, blue] table { 
      0.4 0.9
    };
    \addplot[only marks, mark=text, text mark = {2}, blue] table { 
      0.3 0.8
    };
    \addplot[only marks, mark=text, text mark = {3}, red] table { 
      0.45 0.5
    };
    \addplot[only marks, mark=text, text mark = {4}, red] table { 
      0.6 0.4
    };    
    \addplot[only marks, mark=text, text mark = {5}, red] table { 
      0.5 0.2
    };    
  \end{axis}
};

\draw[->] ($(locations) + (4cm, 2.5cm)$) to [out=45, in=90] (A);

\coordinate(A) at ($(matrix.south) + (0, -3)$);
\draw[fill=lightgray!50!white, rotate around={30:(A)}] (A) ellipse (2cm and 1.25cm);
  \node[] at (A) {
  \footnotesize
%   $\widehat{\mathbf{\Sigma}}$
    $\begin{pmatrix} 
      \sigma_x^2 & \rho\sigma_x\sigma_y \\
      \rho\sigma_x\sigma_y & \sigma_y^2
    \end{pmatrix}$
  };

% down arrows
\coordinate(A) at (-2,-15);
\draw[->] ($(A)+(1,1)$) to [out=270, in=90] node[midway, right, xshift=0.5cm] {locate genetic ancestors} (A) ;
\draw[->] ($(locations) + (4cm, 0.5cm)$) to [out=315, in=90] (A);

%%%%%%%%%%%%
% locate ancestors
%%%%%%%%%%%%%

% red tree
\begin{scope}[xshift = 6cm, yshift = -13cm]

  % sample nodes
  \foreach \n in {3,...,5}{
    \coordinate (n\n) at (\n - 3, -4.5); % define coordinate
    \node[below = 0cm of n\n] {\n}; %node label
  }
  % ancestor node 1
  \foreach \t in {-4}{
    \foreach \n in {4,5}{
      \draw[ultra thick, red] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ultra thick, red] (n4 |- 0,\t) -- (n5 |- 0,\t) coordinate[midway] (n6);
  }    
  % ancestor node 3
  \foreach \t in {-3}{
    \foreach \n in {3,6}{
      \draw[ultra thick, red] (n\n) -- (n\n |- 0,\t);
    }
  \draw[ultra thick, red] (n3 |- 0,\t) -- (n6 |- 0,\t) coordinate[midway] (n8);
  }
  % red mrca stem
  \draw[ultra thick, red] (n8) -- (n8 |- 0,-2.75); 
  % focal ancestor
  \draw[ultra thick, red] (n6 |- 0,-3.7) -- (n6 |- 0,-3) node[midway, right, red] {$t_A$};  
  \draw[fill=black] (n6 |- 0,-3.7) circle (0.15cm) node[right] {$A$};

\end{scope}

% shared times
  \node[anchor = north] (matrix) at (3,-15.75) {
   $\left(\begin{array}{>{\columncolor{red!20}}cccc}
        \rowcolor{red!20}
        \color{red}{t_A} & \color{red}{0} & \color{red}{t_A} & \color{red}{t_A} \\ 
        \color{red}{0} & \color{red}{t_3} & \color{red}{0} & \color{red}{0} \\ 
        \color{red}{t_A} & \color{red}{0} & \color{red}{t_3} & \color{red}{t_6} \\ 
        \color{red}{t_A} & \color{red}{0} & \color{red}{t_6} & \color{red}{t_3} 
  \end{array}\right)$
  };

% ancestor plot
\node[anchor=east] at ($(matrix.west) + (-4.75, -1.75)$) (locations) {
  \begin{axis}[
      cycle list name=black white,
      height=5cm,
      width=5cm,
      xlabel=$x$,
      ylabel=$y$,
      samples=5,
      domain=0:1,
      ymin = 0,
      ymax = 1,
      ticks=none,
      x label style={at={(axis description cs:0.5,0)}, anchor=north, color=gray},
      y label style={at={(axis description cs:0,0.5)},rotate=-90, anchor=east, color=gray}
  ]
%  \addplot+[only marks]  {rand}; %plot some random locations
    \addplot[only marks, mark=text, text mark = {1}, blue] table { 
      0.4 0.9
    };
    \addplot[only marks, mark=text, text mark = {2}, blue] table { 
      0.3 0.8
    };
    \addplot[only marks, mark=text, text mark = {3}, red] table { 
      0.45 0.5
    };
    \addplot[only marks, mark=text, text mark = {4}, red] table { 
      0.6 0.4
    };    
    \addplot[only marks, mark=text, text mark = {5}, red] table { 
      0.5 0.2
    };    
    \addplot[mark=text, text mark = {A}, black] table { 
      0.525 0.375
    };
%    \addplot[domain=0:360,data cs=polar, samples=200, line width=0.7pt, red] (x,{1/(sqrt(1 - 2*cos(x)*cos(x)))});
\addplot [domain=-pi:pi, samples=200, black, line width=0.7pt, rotate around={35:(axis cs:0.525,0.375)}] ({0.525 + sqrt(0.005)*sin(deg(x))}, {0.375 + sqrt(0.008)*cos(deg(x))});
  \end{axis}
};

%arrows
\draw[->] (3,-4) to [out=0, in=90, looseness=1] ($(matrix) + (4,2)$);
\draw[->] ($(matrix.east) + (0.75,0)$) -- (matrix.east) node[midway, below, yshift=-1.25cm] {calculate shared times with ancestor};
\draw[->] (matrix.west) -- ($(matrix.west) + (-1,0)$);

\end{tikzpicture}
\end{center}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\caption{
(Continued on the following page.)
}
\label{fig:overview}
\end{figure}
%\clearpage
\begin{figure}[!t]
\contcaption{
\textbf{Conceptual overview of the approach.}
From a sequence of trees covering the full genome, we downsample to trees at approximately unlinked loci. To avoid the influence of strongly non-Brownian dynamics at deeper times (e.g., glacial refugia, boundaries), we ignore times deeper than $T$, which divides each tree into multiple subtrees (here, blue and red subtrees at locus $i$). From these subtrees we extract the shared times of each lineage with all others. In practice (but not shown here), we use multiple samples of the tree at a given locus, for importance sampling, and also extract the coalescence times for importance sample weights. Under Brownian motion, the shared times describe the covariance we expect to see in the locations of our samples, and so using the times and locations we can find the maximum likelihood dispersal rate (a 2x2 covariance matrix). While we can estimate a dispersal rate at each locus, a strength of our approach is that we combine information across many loci, by multiplying likelihoods, to estimate a single genome-wide dispersal rate. Finally, we locate a genetic ancestor at a particular locus (a point on a tree, here $A$) by first calculating the time this ancestor shares with each of the samples in its subtree, and then using the shared times and dispersal rate to calculate the probability distribution of the ancestor's location conditioned on the sample locations. In practice (but not shown here), we calculate the location of the ancestor of a given sample at a given time across many loci, combining information across loci into a distribution of genome-wide ancestry across space.
}% Continued caption
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

The rate of dispersal, which determines the average distance between parents and offspring, is a key parameter in ecology and evolution. To estimate this parameter we assume that in each generation the displacement of an offspring from its mother is normally distributed with a mean of 0 in each dimension and covariance matrix $\mathbf{\Sigma}$. In two dimensions the covariance matrix is determined by the standard deviations, $\sigma_{x}$ and $\sigma_{y}$, along the $x$ (longitudinal) and $y$ (latitudinal) axes, respectively, as well as the correlation, $\rho$, in displacements between these two axes. The average distance between a mother and its offspring is then $\sqrt{2/\pi} \sigma_i$ in each dimension.

Given this model, the path of a lineage from its ancestral location to the present day location is described by a Brownian motion with rate $\mathbf{\Sigma}$. We therefore refer to $\mathbf{\Sigma}$ as the dispersal rate, which describes the rate of increase in the variance of a location along a single lineage. Lineages covary in their locations because of shared evolutionary histories -- lineages with a more recent common ancestor covary more. Given a tree at a locus we can calculate the covariance matrix of shared evolutionary times and compute the likelihood of the dispersal rate, $\mathbf{\Sigma}$, which is normally distributed given this covariance matrix. At each locus we can estimate the likelihood of the dispersal rate given the tree at that locus allowing us to multiply likelihoods across loci to derive a genome-wide likelihood, and thus a genome-wide maximum likelihood dispersal rate. While the genome-wide dispersal estimate is not biased by correlations between the trees, we simply use every $n^{\mathrm{th}}$ locus across the genome to reduce the redundancy of information we estimate dispersal from.

Under this same model we can also estimate the locations of genetic ancestors at a locus. Any point along a tree at any locus is a genetic ancestor of one or more current day samples. This ancestor's lineage has dispersed away from the location of the most recent common ancestor of the samples, and covaries with current day samples in their geographic location to the extent that it shares times in the tree with them. Under this model, the location of an ancestor is influenced by the locations of all samples in the same tree, including those that are not direct descendants \citep[cf.][]{wohns2021unified}. For example, in Figure \ref{fig:overview} the ancestor's location is not the midpoint of its two descendants (samples 4 and 5); the ancestor's location is also pulled towards sample 3 since the ancestor and sample 3 both arose from a common ancestor. Conditioning on the sample locations, and given the shared times and previously inferred dispersal rate, we can compute the probability the ancestor was at any location, which again is a normal distribution. In contrast to dispersal, for ancestral locations we do not want to multiply likelihoods across loci since the ancestors at distant loci are likely distinct. Instead we calculate the maximum likelihood location of the genetic ancestor at each sparsely-sampled locus to get a cloud of likely ancestral locations genome-wide, and use these clouds to visualize the spatial spread of genetic ancestry backwards through time.

We estimate marginal trees along the genome using \texttt{Relate} \citep{speidel2019method}. While we could in principle estimate trees independently at regions across the genome, a benefit of \texttt{Relate} and similar methods is that they estimate each tree using information from nearby regions. \texttt{Relate} infers a sequence of tree topologies and associated branch lengths, and can return a set of posterior draws of the branch lengths on a given tree \citep[it was the only method that did so for a large number of samples when we began this work, but now see][]{wohns2021unified,deng2024robust}. This posterior distribution of branch lengths is useful to us as the shared times in the tree are key to the amount of time that individual lineages have had to disperse away from one another and we wish to include uncertainty in the times into our method. \texttt{Relate} gives us a posterior distribution of branch lengths that is estimated using a coalescent prior, which assumes a panmictic population of varying population size (the size changes are estimated as part of the method), where any two lineages are equally likely to coalesce. This panmictic prior results in a bias in the coalescent times under a spatial model, where geographically proximate samples are more likely to coalesce. To correct for this bias we make use of importance sampling to weight the samples of branch lengths at each locus. We then calculate the weighted average likelihood over our draws of our sample of trees at a locus (or loci), so that it is as if they were drawn from a prior of branching Brownian motion \citep{meligkotsidou2007postprocessing}. 

In practice we concentrate on the recent past history of our sample. For our estimates of dispersal rates in particular we do not want to assume that our model of Gaussian dispersal (and branching Brownian motion) holds deep into the past history of the sample. This is because the long-term movement of lineages is constrained by geographic barriers (e.g., oceans) and larger scale population movements may erase geographic signals over deep time scales. On a theoretical level ignoring the deep past may also be justified because in a finite habitat the locations of coalescence events further back in time become independent of sampling locations as lineages have moved around sufficiently \citep{wilkins2002coalescent}. Thus we only use this geographic model to some time point in the past ($T$), and at each locus we use the covariance of shared branch lengths based on the set of subtrees formed by cutting off the full tree $T$ generations back. 

Branching Brownian motion, also known as the Brownian-Yule process, is a simple model of spatial genealogies in a continuous population \citep{malecot1948mathematics,wright1943isolation}. This simplicity comes at a cost: the lack of local density-dependence generates non-uniform population density \citep{felsenstein1975pain} and the assumption of unbounded space overestimates the distance between samples \citep{kalkauskas2021sampling}. More complex models, such as the spatial Lambda Fleming-Viot process \citep{barton2010new}, overcome these issues and more, but are computationally expensive for inference \citep{wirtz2023connections}. Branching Brownian motion remains an analytically tractable model and a reasonable approximation over short-time scales \citep{edwards1970estimation,rannala1996probability,meligkotsidou2007postprocessing,novembre2009likelihood}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Simulations}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

We first wanted to test the performance of our method in a situation where the true answers were known. To do this we used a combination of spatially-explicit forward-time simulations \citep{haller2023slim}, coalescent simulations \citep{kelleher2016efficient}, and tree-sequence tools \citep{haller2019tree,kelleher2019inferring,speidel2019method} to compare our estimates of dispersal rates and ancestor locations with the truth (see \nameref{sec:methods}). This was also an opportunity to compare our estimates using the true trees vs.\ the \texttt{Relate}-inferred trees, to examine the influence of uncertainty in tree inference.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Dispersal rates}

Our method systematically underestimates simulated dispersal rates (Figure \ref{fig:sims}A), as expected given that our simple unbounded Brownian motion model allows the samples to be more broadly distributed than the finite habitat \citep[e.g.,][]{ianni2023exploring,kalkauskas2021sampling}. Ignoring the distant past tends to reduce this underestimate (see $T=1000$ in Figure \ref{fig:sims}A), but ignoring too much of the past increases noise and can lead to even larger underestimates (see $T=100$ in Figure \ref{fig:sims}A). Despite the fact that we tend to underestimate the simulated dispersal rate, our estimates are highly correlated with the simulated values and we can interpret the dispersal rate inferred from the true trees as a true `effective' dispersal rate given the habitat boundaries, local competition, etc. 

Encouragingly, the dispersal estimates from the inferred trees are highly correlated with those from the true trees, with a slight upward bias. This upward bias can be explained by isolation-by-distance and errors in inferred tree topologies, causing geographically more distant samples to be mistakenly inferred as closer relatives. The upward bias may also result from \texttt{Relate} tending to underestimate longer coalescence times \citep{brandt2022evaluation}, which is consistent with the bias decreasing at smaller cutoff times. The bias increases as we use sample fewer trees at each locus (Figure \ref{fig:sup_sigma_imp}), showing that our spatial prior implemented via importance sampling improves our inference. Sampling more individuals reduces noise across replicates but can actually increase bias (Figure \ref{fig:sup_sigma_samples}), perhaps because of a greater chance of an error when inferring a larger tree. More work is needed to determine biases in ancestral recombination graphs inferred from spatial data more generally, and to find additional ways to overcome this bias, e.g., using only the most informative trees.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%simulation figure%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!b]
%\captionsetup{labelformat=empty}
\begin{center}
\begin{tikzpicture}[remember picture]

  % disp rates, no cutoff
  \node[] (Tnone) {
    \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone.pdf}    
  };
  \node[left = 0cm of Tnone.north west, anchor = north west, yshift = -0.15cm] (A) {
    \textbf{A}
  };
  \node[above = -0.75cm of Tnone, anchor=south] () {
    \footnotesize{$T = \infty$}
  };

  % disp rates, 1000 cutoff
  \node[right = -0.2cm of Tnone] (T1000) {
    \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000.pdf}
  };
  % \node[left = 0cm of T1000.north west, anchor = north west, yshift = -0.15cm] (B) {
  %   \textbf{B}
  % };
  \node[above = -0.75cm of T1000, anchor=south] () {
    \footnotesize{$T = 1000$}
  };

  % disp rates, 100 cutoff
  \node[right = -0.2cm of T1000] (T100) {
    \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100.pdf}
  };
  % \node[left = 0cm of T100.north west, anchor = north west, yshift = -0.15cm] (C) {
  %   \textbf{C}
  % };
  \node[above = -0.75cm of T100, anchor=south] () {
    \footnotesize{$T = 100$}
  };

  % single lineages
  \node[below = -0.6cm of Tnone.south west, anchor = north west] (CI-1) {
    \includegraphics[height=0.325\linewidth, trim={2.5cm 0cm 3.3cm 0cm}, clip]{single-lineages-100t.pdf}
  };
  \node[right = -0.15cm of CI-1] (CI-2) {
    \includegraphics[height=0.325\linewidth, trim={3.85cm 0cm 3.3cm 0cm}, clip]{single-lineages-1000t.pdf}
  };
  % \node[right = 0cm of CI-2] (CI-3) {
  %   \includegraphics[height=0.25\linewidth, trim={3.75cm 0cm 3.3cm 0cm}, clip]{single-lineages-1000t.pdf}
  % };  
  \node[above left = 0cm of CI-1.north west, anchor = north west, yshift = 0.05cm] (E) {
    \textbf{B}
  };

  % RMSE single lineages
  \node[right = -0.55cm of CI-2] (RMSE) {
    \includegraphics[height=0.325\linewidth, trim = {0cm 0 0cm 0cm}, clip]{rmse-single-lineages.pdf}
  };
  \node[left = -0.25cm of RMSE.north west, anchor = north west, yshift = -0.05cm] (D) {
    \textbf{C}
  };  

  % ancestor clouds
  \node[below = -0.5cm of CI-1] (BLUP-1) {
    \includegraphics[height=0.325\linewidth, trim={2.5cm 0cm 3.3cm 0cm}, clip]{ancestor-clouds-100t.pdf}
  };
  \node[below = -0.5cm of CI-2] (BLUP-2) {
    \includegraphics[height=0.325\linewidth, trim={3.85cm 0cm 3.3cm 0cm}, clip]{ancestor-clouds-1000t.pdf}
  };
  % \node[below = -0.5cm of CI-3] (BLUP-3) {
  %   \includegraphics[height=0.25\linewidth, trim = {3cm 0 0cm 0}, clip]{ancestor-clouds-1000t.pdf}
  % };  
  \node[above left = 0cm of BLUP-1.north west, anchor = north west, yshift = 0.05cm] (G) {
    \textbf{D}
  };

  % RMSE clouds
  \node[below = -0.5cm of RMSE] (RMSE-mean) {
    \includegraphics[height=0.325\linewidth, trim = {0cm 0 0 0cm}, clip]{rmse-ancestor-clouds.pdf}
  };
  \node[left = -0.25cm of RMSE-mean.north west, anchor = north west, yshift = -0.05cm] (F) {
    \textbf{E}
  };  

\end{tikzpicture}  
\end{center}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\caption{}
%\end{figure}
%%\clearpage
%\begin{figure}[!t]
%  \ContinuedFloat
%  \caption{
%}
%\label{fig:sims}
%\end{figure}
\caption{
(Continued on the following page.)
}
\label{fig:sims}
\end{figure}
%\clearpage
\begin{figure}[!t]
\contcaption{
\textbf{Simulations.}
\textbf{(A) Accuracy of genome-wide dispersal rates.} 
Maximum composite likelihood estimates of dispersal rate (in `x' and `y' dimensions) using the true trees vs.\ \texttt{Relate}-inferred trees for three different time cutoffs, $T$. Colours indicate the simulated dispersal rates, which are given by the corresponding lines. 
\textbf{(B) Locating genetic ancestors at a particular locus.}
95\% confidence ellipses for the locations of genetic ancestors for three samples at a single locus (using the true trees and the simulated dispersal rate). The `o's are the sample locations and the `x's are the true ancestral locations.
\textbf{(C) Accuracy of locating genetic ancestors at individual loci.}
Root mean squared errors between the true locations of ancestors and the mean location of the samples (red), the current locations of the samples (green), and the maximum likelihood estimates from the inferred (orange) and true (blue) trees.
\textbf{(D) Locating genetic ancestors at many loci.}
Contour plots of the most likely (using the true trees; blue) and the true (gray) locations of genetic ancestors at every 100th locus for a given sample.
\textbf{(E) Accuracy of mean genetic ancestor locations.}
Root mean squared errors between the true mean location of genetic ancestors and the mean location of the samples (red), the current locations of the samples (green), and the mean maximum likelihood estimates from the inferred (orange) and true (blue) trees. To reduce computation time we only attempt to locate the first 10 samples. 
In all panels, there are 10 replicate simulations for each combination of time cutoff and dispersal rate. We sample 50 diploid individuals at random and use every 100th locus, with 1000 importance samples at each. Panels \textbf{B}-\textbf{E} have no time cutoff, $T=\infty$, and were simulated with a dispersal rate given by green lines in panel \textbf{A}. In panels \textbf{C} and \textbf{E}, the inferred-tree ancestor location estimates use the inferred-tree dispersal estimates. 
}% Continued caption
\end{figure}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Locating ancestors}

We next wanted to test our ability to locate the genetic ancestor of a sampled genome at a given locus and a given time. With a single tree, our likelihood-based method gives both a point estimate (maximum likelihood estimate, MLE) and a 95\% confidence ellipse (under the Brownian motion model), with only the latter relying on a genome-wide dispersal rate. With more than one tree we importance sample over likelihoods, numerically finding the maximum, which depends on the dispersal rate. We also have developed a best linear unbiased predictor (BLUP) of ancestral locations -- importance sampling over analytically calculated MLE locations -- that is faster to calculate, makes fewer assumptions, and is independent of the dispersal rate. Here we focus on the full importance-sampled likelihood method for estimating ancestral locations, but the BLUP method is also implemented in our software and performs very similarly (Figure \ref{fig:blups}).

Figure \ref{fig:sims}B shows the 95\% confidence ellipses for the locations of the ancestors of three samples at one particular locus at two different times in the past, using the true trees and the simulated dispersal rate. While the ellipses generally do a good job of capturing true ancestral locations (the `x's), the size of an ellipse at any one locus grows relatively rapidly as we move back in time, meaning that at deeper times (or higher dispersal rates) any one locus contains little information about an ancestor's location \citep[as is the case for ancestral state reconstruction in phylogenetics;][]{schluter1997likelihood}.

Figure \ref{fig:sims}C shows the resulting error in the MLE ancestor locations, using the true or inferred trees, and compares this to sensible straw-man estimates (the current location of each sample and the mean location across samples). While our method outperforms the straw-man estimates, the mean squared error in our inferred location of the ancestor at a locus grows relatively rapidly back in time, as expected under Brownian motion.

Given the large uncertainty of an ancestor's location at any one locus, we combine information across loci and consider a cloud of MLE ancestor locations from loci across the genome for a particular sample at a particular time in the past (Figure \ref{fig:sims}D). The genome-wide mean of the MLE locations remains closer to the true mean location of genetic ancestors (Figure \ref{fig:sims}E) than at any one locus, even with the inferred trees, suggesting our method can successfully trace major trends in the geographic ancestry of a sample deeper into the past. The error in ancestral locations appears relatively robust to the number of trees sampled per locus (Figure \ref{fig:anc-locs-M}) and the number of samples (Figure \ref{fig:anc-locs-k}).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Empirical application: \textit{Arabidopsis thaliana}}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\textit{A. thaliana} has a complex, and not yet fully resolved, spatial history \citep{fulgione2018archaic,hsu2019postglacial}, including range expansions, admixture between multiple glacial refugia, and long-distance colonization.
To further examine this history, we originally applied our method to 1135 \textit{Arabidopsis thaliana} accessions from a wide geographic range \citep{alonso2016}.
Unfortunately, the dispersal rates we estimated from this data were unreasonably large, on the order of $10^4$ km$^2$/generation, even after removing pairs of near-identical samples and geographic outliers \citep[for more details see our preprint,][]{osmondcoop2021}.
We next analyzed an even larger dataset of roughly 1500 individuals from a broader geographic extent \citep{durvasula2017african}, but estimated a similarly large dispersal rate (results not shown but full pipeline available at \url{https://github.com/mmosmond/spacetrees-ms}).
The very high dispersal rate estimates from these datasets is an artefact of long runs of identity between pairs of geographic separated sequences \citep[see Figure S6 of][]{osmondcoop2021}. We believe that these stretches of identity result from the incorrect imputation of missing genotypes, a problem that could not be resolved through various data filters. Imputation was required to infer the genealogies as \texttt{Relate} does not allow sites with missing genotypes. However, imputation can obscure rare alleles, which artificially lowers the divergence between similar sequences. This lowers coalescence times and biases dispersal estimates upward. While it may be possible to improve imputation, we resolved this issue by analyzing a smaller dataset with 66 long-read genomes \citep{wlodzimierz2023cycles}, where we could avoid imputing and capture more genetic variation \citep{igolkina2024towards} and therefore be much more confident in the resulting genealogies. 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{\textit{A.\ thaliana} genealogies}

From 66 haploid genomes we used \texttt{Relate} \citep{speidel2019method} to infer the genome-wide genealogy (87518 trees), estimate effective population sizes through time (Figure \ref{fig:nes}), and resample branch lengths for importance sampling (see \nameref{sec:methods}). The genealogies are publicly available at \url{https://doi.org/10.5281/zenodo.11456353}, which we hope will facilitate additional analyses \citep[e.g., inferring selection;][]{stern2019approximate,stern2020disentangling}.

\textit{A. thaliana} is a selfer with a relatively low rate of outcrossing \citep{bomblies2010local,platt2010scale}, thus it is worth taking a moment to consider the impact of selfing on our inferences. We chose \textit{A. thaliana} because of its large sample size, broad geographic sampling, and intriguing spatial history. Further, the availability of inbred accessions means that many samples have been well-studied and phasing is relatively straightforward. On the other hand, the high rate of selfing lowers the effective recombination rate and so is expected to increase the correlation in genealogies along the genome \citep{nordborg2000linkage}. However, in practice, linkage disequilibrium breaks down relatively rapidly in \textit{A. thaliana}, on the scales of tens of kilobases \citep{kim2007recombination}, such that many trees along the genome should have relatively independent genealogies. A related issue is that the individuals with recent inbreeding (selfing) in their family tree will have fewer genealogical and genetic ancestors than outbred individuals. Thus in any recent time-slice there are a reduced number of independent genetic ancestors of a individual from a selfing population, but even with relatively low rates of outbreeding the number of ancestors still grows rapidly \citep{lachance2009inbreeding}, meaning that many trees along the genome should have relatively independent spatial histories. Finally, while the effective recombination rate may vary through time along with rates of selfing, \texttt{Relate} uses a mutational clock to estimate branch lengths, and thus they should be well calibrated to a generational time scale.

In our analyses below, we use every 100th tree starting from the beginning of each autosome, for a total of 878 trees (hereafter, loci). At each locus we resample the branch lengths 1000 times, giving us 1000 importance samples (hereafter, trees) at each locus.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Estimating dispersal}

We first used the trees at all sampled loci to estimate dispersal rate (Figure \ref{fig:thal_sigma_displaced}, inset). Using the full trees, back to the most recent common ancestor of each, we estimate a dispersal rate of about $30$ km$^2$/generation. 

To check that this dispersal estimate is well calibrated we can compare it to simpler estimates. To do this we first calculated twice the average pairwise coalescent time over all loci (using just 1 tree per locus) for each pair of samples, $t_{ij}$ (multiplying $t_{ij}$ by the mutation rate then gives the expected pairwise nucleotide diversity \citep{ralph2020efficiently}). A simple estimate of the dispersal rate is then the slope of the regression of the squared pairwise geographic distances, $d_{ij}^2$, on the average coalescent times, $t_{ij}$. This gives a dispersal estimate of roughly 10 km$^2$/generation, as does taking the average of $d_{ij}^2/t_{ij}$ over pairs \citep{ianni2023exploring}. Replacing the pairwise coalescent times, $t_{ij}$, with pairwise nucleotide diversity divided by the mutation rate ($\pi_{ij}/(7\times10^{-9})$) gives a near identical estimate that is independent of our inferred trees. Our tree sequence estimate is expected to be larger than these simpler estimates because taking the arithmetic mean coalesecnt times over trees will lessen the amplifying effect of those trees with shorter coalescent times. Using instead the harmonic mean pairwise coalescent time over trees, we estimate a dispersal rate of roughly 40 km$^2$/generation, similar to our estimate. 

We next examined the effect of cutting off our inference deeper in time (Figure \ref{fig:thal_sigma_displaced}, inset). Cutting the trees off $10^6$ generations ago has essentially no effect, as it removes few coalescent events. With cutoffs of $10^5$ and $10^4$, the estimated dispersal rate increases to about 35 and 60 km$^2$/generation, respectively, which we believe is because finite habitat boundaries lower the effective dispersal rate over longer timescales but also suggests faster rates of movement post-glaciation. More recent cutoffs leave very few coalescent events and hence little information. We cannot reliably apply the same cutoffs to the simpler pairwise dispersal estimates discussed above because there are few pairs of samples that have an average genetic distance below the cutoffs, highlighting the increased amount of information in our method.

These dispersal estimates are considerably higher than a recent estimate, using machine learning directly on genotype data, that was on the order of 1 km$^2$/generation \citep{smith2023dispersal}. This previous study used a similar sample size but a much narrower geographic range of more closely-related individuals. It is therefore likely that this lower estimate more strongly reflects shorter-term local dispersal, while we are striking a balance between these slower recent movements and faster movements that have occasionally occurred in deeper time. To get a sense of how large a dispersal rate we should expect, `non-relict' lineages are proposed to have dispersed from the Black Sea to the Atlantic Ocean (about 2500 km) in the past $10^4$ generations \citep{lee2017post,hsu2019postglacial}. This suggests some lineages should have $d_i^2/t_i$ values of at least 300 km$^2$/generation. We therefore think our effective dispersal rate is a reasonable one given the Brownian motion model and hypothesized movements in \textit{A. thaliana}.

We estimate slightly faster dispersal along latitude than along longitude, especially in the past $10^4$ generations. This was confirmed by estimating the dispersal rate separately along latitude and longitude, removing dependence on the sample locations (see methods). Faster dispersal along latitude in the past $10^4$ generations is consistent with rapid post-glacial expansion. Previous studies have instead emphasized a rapid westward expansion of the non-relict lineages from an eastern glacial refugium, facilitated by relatively weak environmental gradients and, potentially, human movements and disturbance \citep{alonso2016,lee2017post,hsu2019postglacial}. It is possible that our longitudinal dispersal estimate would increase substantially if we included non-relict samples from further east.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]

\begin{center}

\begin{tikzpicture}

  \node[] (A) {
    \includegraphics[width=\textwidth, trim=40 10 40 10, clip]{mean-displacements.pdf}
  };

  % \node[yshift = -55, xshift = 42] (B) at (A) {
  \node[yshift=-80] (B) at (A) {
    % \includegraphics[width=0.55\textwidth]{{dispersal-rates-dark.pdf}}
    \includegraphics[width=0.55\textwidth]{{dispersal-rates.pdf}}
  };
  
\end{tikzpicture}  

\end{center}

\caption{
\textbf{Dispersal rates (inset) and major trends in geographic ancestries.}
Vectors start at sample locations (circles) and point to the mean location of ancestors across 878 loci $10^4$ generations ago. Samples colored by genomic principle component group \citep{wlodzimierz2023cycles}.
}

\label{fig:thal_sigma_displaced}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Visualizing major trends in geographic ancestries}

We next used the trees at all sampled loci and our dispersal estimate to locate genetic ancestors. Our method provides an estimate of the location of the ancestor at every locus for every sample back through time, giving a rich resource for visualizing the geography of genetic ancestry. As a first step, we visualize the mean ancestral location for every sample at a given time, averaged over loci, to visualize major geographic trends and detect samples with unusual geographic ancestries (Figure \ref{fig:thal_sigma_displaced}).
When locating ancestors we use the full trees, which relate every sample to every other, but use the dispersal estimate from a time cutoff of $10^4$ generations and only locate ancestors back to that cutoff time. 

Consistent with a rapid recent expansion of the non-relict ancestry \citep{lee2017post,hsu2019postglacial}, lineages from the `Eurasian' PCA group (hereafter, non-relicts) move relatively quickly towards each other when looking backwards in time, a visual display of their rapid coalescence over wide geographic distances. Looking forward through time (i.e., reversing the direction of the arrows), this rapid geographic expansion of non-relict ancestry is especially pronounced further north, as might be expected with retreating glaciers and more recent human disturbance \citep{lee2017post}. The fact that the non-relict lineages are all pulled towards the bulk of non-relict samples, in southern France, highlights the limitations of sampling and methods based on contemporary data alone. The non-relicts likely spread from further east, with a refugium perhaps near the Black Sea \citep{lee2017post,hsu2019postglacial}, but this dataset does not have enough samples in that area to infer this.

Meanwhile nearly all lineages from the `Iberian relict' PCA group (hereafter, Iberian relicts) move relatively little, and towards themselves rather than towards the non-relict ancestors. This is a visual representation of the fact that much of the ancestry of these Iberian relicts has a very different geographic history than much of the non-relicts' ancestry \citep{alonso2016,fulgione2018archaic}. There are, however, two exceptions, with Iberian relicts Met-6 and Evs-12 having (overlapping) mean displacements that are much more typical of a non-relict. We explore these outliers more below.

Finally, we see that lineages from the `non-Iberian relict' PCA group (hereafter, non-Iberian relicts) move towards the other samples, slightly more towards the Iberian relicts than the non-relicts. This movement is, however, relatively slow, emphasizing the known deep divergence of each of these samples with any other \citep{alonso2016,fulgione2018madeiran,fulgione2018archaic,fulgione2022parallel}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Visualizing the geographic sources of ancestry}

We can also look at variation across loci in the geographic ancestry of a sample. We illustrate this while investigating the outliers above, the Iberian relicts with mean displacements more typical of a non-relict, Met-6 and Evs-12. In particular, we draw great circles connecting a sample's location to that of each of its ancestors at a given time, and include a histogram of the direction to the ancestors (a ``windrose" plot). 

Figure \ref{fig:thal_windrose} compares four samples from Spain. The first is a typical non-relict (9883) whose ancestors are nearly all located to the northeast, as expected given the locations of the other non-relict samples and the proposed expansion from the east. The second is a typical Iberian relict (Hum-2) whose ancestors are mostly located just to the south of it, near the center of the Iberian relict samples, as expected given the divergence between non-relicts and relicts and the longer history of relicts in the region. This sample also has a substantial fraction of ancestors to the northeast, likely illustrating some recent admixture from the non-relict ancestry. The third and fourth samples are the outlier Iberian relicts (Met-6 and Evs-12), who both have many ancestors that are located relatively far to the northeast, much like the non-relict. This likely reflects quite extensive admixture between Iberian relicts and non-relicts \citep{alonso2016,fulgione2018archaic}, which can also be seen in the fact that the genealogical nearest neighbour proportions \citep{kelleher2019inferring} for these two samples are heavily weighted towards non-relicts (at roughly 75\% and 83\%, while the mean across all Iberian relicts is roughly 30\%).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{figure}[!htb]

\begin{center}

\begin{tikzpicture}

 \node[] (A) {
    \includegraphics[width=0.5\textwidth]{windrose14.pdf}
  };
 \node[below = -0.25cm of A, fill=black, text=white, anchor=center] {9883};
 \node[above = 0cm of A, text=tab10_blue, anchor=center] {non-relict};
 
 \node[right = 0cm of A] (B) {
    \includegraphics[width=0.5\textwidth]{windrose39.pdf}
  };
 \node[below = -0.25cm of B, fill=black, text=white, anchor=center] {Hum-2};
 \node[above = 0cm of B, text=tab10_orange, anchor=center] {Iberian relict};

 \node[below = 0cm of A] (C) {
    \includegraphics[width=0.5\textwidth]{windrose59.pdf}
  };
 \node[below = -0.5cm of C, fill=black, text=white, anchor=center] {Met-6};

 \node[right = 0cm of C] (D) {
    \includegraphics[width=0.5\textwidth]{windrose43.pdf}
  };
 \node[below = -0.5cm of D, fill=black, text=white, anchor=center] {Evs-12};

\end{tikzpicture}  

\end{center}

\caption{
\textbf{Visualizing the geographic sources of ancestry.}
Great circles connecting sample and ancestor locations at 878 loci $10^4$ generations ago. Polar histograms (`windroses') show the distribution of direction in ancestral locations from the sample. Non-relicts in blue, Iberian relicts in orange.
}

\label{fig:thal_windrose}
\end{figure}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Visualizing the movement of geographic ancestries over time}

We next looked at the entire distribution of genetic ancestor locations, over loci, for a given sample at a few given times (Figure \ref{fig:thal_clouds}). We see that the ancestors of the most northern sample (non-relict 9336) quickly move southwest towards the other non-relicts, illustrating a rapid northward expansion post-glaciation. In stark comparison, the ancestors of the most diverged sample (non-Iberian relict 22005, in Madeira) move relatively little in the past $10^4$ generations, consistent with estimates that \textit{A. thalaina} colonized Madeira roughly 70 thousand years ago \citep{fulgione2018madeiran}.

There is a slight bimodality in the locations of the ancestors of the northern non-relict sample (9336) at later times. This bimodality also exists in the locations of the ancestors of the other sample from Sweden (non-relict 6137, not shown). It is tempting to wonder if this bimodality is a visualization of multiple pulses of northern expansions. In particular, based on the presence of abnormal amounts of relict ancestry in some northern samples, it has been hypothesized that relict lineages were present across much of Europe before being largely replaced, especially at midlatitudes, by a westward expansion of non-relict lineages \citep{lee2017post,hsu2019postglacial}. These non-relict lineages are associated more with disturbed areas and so are expected to have arrived in the north more recently, and in lower numbers, explaining why some relict ancestry remains there. We therefore checked to see if the bimodality in ancestor locations of the Swedish samples could be explained by the placement of these focal samples in the corresponding trees. There was, however, no clear correlation between ancestor location and the proportion of genealogical nearest neighbours that were non-relict vs.\ Iberian relict. It of course remains possible there were multiple bouts of colonization, perhaps even from within a given PCA group, that could be visualized with other metrics.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% geographic ancestries over time
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]

\begin{center}

\begin{tikzpicture}

  \node[] (A) {
    \includegraphics[width=\textwidth]{sample12-cloud.pdf}
  };
  \node[left = -0.25cm of A, fill=black, text=white, anchor=center, rotate=90] {9336};
  \node[above = 0cm of A, text=tab10_blue, anchor=center, xshift=-4.5cm] {non-relict};

  \node[below = -0.1cm of A] (B) {
    \includegraphics[width=\textwidth]{sample42-cloud.pdf}
  };
  \node[left = -0.35cm of B, fill=black, text=white, anchor=center, rotate=90] {22005};
  \node[below = 0cm of B, text=tab10_green, anchor=center, xshift=-4.5cm] {non-Iberian relict};

  \node[below = 0cm of A, anchor=center, xshift=-4.5cm, fill=black, text=white] {$10^3$ generations ago};
  \node[below = 0cm of A, anchor=center, fill=black, text=white] {$5000$ generations ago};
  \node[below = 0cm of A, anchor=center, xshift=4.5cm, fill=black, text=white] {$10^4$ generations ago};

  
\end{tikzpicture}  

\end{center}

\caption{
\textbf{Visualizing the movement of geographic ancestries over time.}
Ancestor locations at 878 locations (circles) and kernel density estimates (contours). Non-relicts in blue, non-Iberian relcits in green.
}

\label{fig:thal_clouds}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Visualizing the geographic ancestries of groups of samples}

The dataset contains samples drawing ancestors from at least two hypothesized glacial refugia, including the non-relcits with ancestry predominately from a refuge near the Black Sea \citep{lee2017post,hsu2019postglacial} and the Iberian relict samples with substantial ancestry from a refuge in northern Africa \citep{alonso2016,durvasula2017african,fulgione2018madeiran}. To visualize these alternative geographic ancestries we located the ancestors of all Iberian relicts and compared that distribution to the distribution of the locations of the ancestors of all non-relicts in mainland Spain (Figure \ref{fig:thal_ancestries}). We see that the two ancestries are largely geographically overlapping in the past $10^3$ generations but then begin to diverge due to the majority of non-relicts ancestors moving northeast. There is also considerable variation in the locations of both ancestries, emphasizing substantial admixture. Having a few samples from north Africa would likely pull many of the ancestors of the Iberian relict samples south, further emphasizing the divergent geographic histories.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%Non-Relict - Admixed  - Relict%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]

\begin{center}

\begin{tikzpicture}

 \node[] (A) {
    \includegraphics[width=0.5\textwidth, trim = 50 50 0 0, clip]{relict-vs-nonrelict-clouds-1.pdf}
  };
 \node[below = 0cm of A.south, anchor=center, fill=black, text=white, xshift=-0.75cm] {$10^3$ generations ago};
 
  \node[right=0cm of A] (B) {
    \includegraphics[width=0.5\textwidth, trim = 50 50 0 0, clip]{relict-vs-nonrelict-clouds-10.pdf}
  };
 \node[below = 0cm of B.south, anchor=center, fill=black, text=white, xshift=-0.75cm] {$10^4$ generations ago};

 \node[right = -0.1cm of A, anchor=center, rotate=90, text=tab10_blue, xshift=10] (nr){non-relict};
 \node[left = 0cm of nr, anchor=east, rotate=90, text=tab10_orange] {Iberian relict};

\end{tikzpicture}  

\end{center}

\caption{
\textbf{Visualizing the geographic ancestries of groups of samples.}
Kernel density estimates (contours and marginal distributions) of the locations of ancestors at 878 loci for all Iberian relicts (orange) and all non-relicts in Spain (blue).
}

\label{fig:thal_ancestries}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Discussion}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Summary of main results}

We have developed a method that uses a sequence of trees along a recombining genome -- a genome-wide genealogy -- to estimate individual-level dispersal rates and locate genetic ancestors (Figure \ref{fig:overview}). At the core of our method is a simple model of Brownian motion, allowing likelihoods to be quickly calculated from shared evolutionary times and sample locations. This also allows us to work in continuous space, negating the need to group individuals into discrete populations. On top of this we layer on importance sampling to correct for bias in inferred branch lengths and add a time cutoff to ignore strong violations of the model in the deep past. Simulation tests show that our method can estimate a meaningful effective dispersal rate and visualize major trends in geographic ancestries hundreds of generations into the past (Figure \ref{fig:sims}).vApplying our method to \textit{Arabidopsis thaliana}, we estimate relatively rapid dispersal since the last glaciation, especially across latitude (Figure \ref{fig:thal_sigma_displaced} inset), and show how locating genetic ancestors allows us to visualize $i)$ major trends in geographic ancestries (Figure \ref{fig:thal_sigma_displaced}), $ii)$ the geographic sources of ancestry (Figure \ref{fig:thal_windrose}), $iii)$ movements of individual geographic ancestries over time (Figure \ref{fig:thal_clouds}), and $iv)$ the geographic ancestries of groups of samples (Figure \ref{fig:thal_ancestries}).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Comparison with previous methods}

% history
The idea of using trees to estimate continuous ancestral characters and their rate of change is an old one. This was originally applied to population-level characters, such as the frequency of genes in a population and their rate of genetic drift, in a phylogenetic context \citep{cavalli1964analysis,cavalli1967phylogenetic,edwards1970estimation,felsenstein1973maximum,felsenstein1985phylogenies,grafen1989phylogenetic}. DNA sequencing later allowed the inference of a single tree relating individual samples for a sufficiently long non-recombining sequence (as found on the human Y chromosome, in a mitochondrial genome, or in the nuclear genome of a predominately non-recombining species), which led to estimates of dispersal rates and ancestral locations under the banner of `phylogeography' \citep{avise2009phylogeography,knowles2009statistical}. Phylogeography has proven incredibly useful, especially to infer the geographic origin and spread of viruses \citep{biek2007high,lemey2009bayesian,lemey2010phylogeography,bedford2010global,volz2013viral}, such as SARS-CoV-2 \citep{worobey2020emergence,lemey2020accommodating,dellicour2021phylodynamic,dellicour2021dispersal,martin2021insights}. Extending phylogeography to frequently-recombining sequences is not straightforward as there are then many true trees that relate the samples. Only recently has it become feasible to infer the sequence of trees, and their branch lengths, along a recombining genome \citep{rasmussen2014genome,speidel2019method,wohns2021unified}.
Our method capitalizes on this advance to use some of the enormous amount of information contained in a tree sequence of a large sample in a recombining species.

% wohns
A related method was demonstrated by \cite{wohns2021unified}, who inferred the geographic location of coalescent nodes in a `succinct' tree sequence \citep{kelleher2018efficient}, where information about nodes and edges are shared between trees. Our approach differs from theirs in a number of ways. First, they utilize the sharing of information about nodes and edges across trees to very efficiently geographically locate every node exactly once, placing each `parent' node at the midpoint between its two `child' nodes locations and iterating up the entire tree sequence simultaneously (rather than up each local tree individually). In addition to being fast, this has the advantage of using information from all the trees in a tree sequence. In contrast, we locate ancestors independently at each local tree we consider. As some ancestors (represented as nodes and edges) are shared between nearby trees along the genome (though we don't know precisely for how long since we lose that information when converting the \texttt{Relate}-inferred genealogy into a tree sequence), we avoid locating the same ancestors multiple times by sparsely sampling the trees (e.g., here we used every 100th tree). (Note that while we choose trees that have low linkage disequilibrium with one another and are therefore essentially unlinked, in the very recent past they will share ancestors but will quickly become independent \citep{wakeley2012gene}.) Our approach therefore uses less of the information in the tree sequence, in this sense. On the other hand, when locating an ancestor we not only use information `below' this ancestor (i.e., its descendants' locations and relations) but also the information `above' the ancestor, due to the ancestor's lineage sharing time and a recent common ancestor with non-descendant lineages. Further advantages of our method include the ability to estimate dispersal rates and uncertainties in ancestor locations (as we have taken a parametric approach), as well as accounting for uncertainty in branch lengths using importance sampling (this could be extended to capture uncertainty in the topologies as well once they can be efficiently sampled).

%ARGs
Ideally we would merge the two methods, to efficiently use information from all the (correlated) trees in the tree sequence and all the information above and below ancestors. Two recent approaches have developed in this direction. One approach \citep{grundler2024geographic} infers a dispersal rate and ancestral locations by placing each node at the location that minimizes a migration cost averaged over the trees in a succint tree sequence that the node appears in. A second approach \citep{deraje2024inferring} extends the branching Brownian motion model to the full ancestral recombination graph \citep[as inferred using a program like \texttt{ARGweaver;}][]{rasmussen2014genome}, where the lineages meet each other forward-in-time at recombination nodes. Niether of these approaches yet account for uncertainty in tree inference.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Future directions}

% alternative likelihood models
We have chosen to use a very simple model of Brownian motion with a constant dispersal rate over time. Informally we view this as a prior on the movement of the lineages, that can be overcome if the trees are informative about long distance dispersal. However, there are a number of extensions that could readily be applied. For instance, we could allow dispersal rates to vary between branches \citep{o2006testing}, e.g., to compare dispersal rates in different parts of a species' range. Or we could model dispersal under a more complex model, like the early burst \citep{harmon2010early} or a L{\'e}vy process \citep{landis2013phylogenetic}, which may help identify periods of range expansion or sudden long-range dispersal. Alternatively we could take a Bayesian approach, allowing much greater flexibility and the ability to incorporate many recent advances in phylogeography. For example, one could then model dispersal as a relaxed random walk \citep{lemey2010phylogeography}, which may be more appropriate for sample locations that are very non-normal, and could incorporate habitat boundaries. Or, given that there is large variance in the inferred locations of distant ancestors at any one locus \citep{schluter1997likelihood} but very many loci, we could take an `empirical Bayes' approach and use the posteriors on ancestral locations over many loci to set a prior for a given locus. This might be especially helpful at deeper times, e.g., tracing human ancestors back hundreds of thousands of years, where the noise in the ancestral location at any one locus is large, yet we can be relatively certain that the majority of lineages are in Africa. We might alternatively set priors to test hypotheses, e.g., if we surmise there were multiple glacial refugia during the last glaciation we can set a prior on ancestral location with peaks at these hypothesized locations and infer what percent of a sample's lineages descended from each. Models of ancestral locations based on past climatic- and ecological-niche models could provide a rich source of data for building such priors, and given the large amounts data available in recombining sequences these models could be subject to rigorous model choice. Finally, it might also be interesting to use machine learning. For example, \texttt{disperseNN} \citep{smith2023dispersal} and \texttt{Locator} \citep{battey2020predicting} use neural networks to estimate dispersal rates and infer the locations of extant individuals, respectively, from unphased genotype data. In essence this means these methods simultaneously determine the relationships between samples and estimate spatial parameters.Separating these two steps by first inferring a tree sequence and then supplying the structure of this tree sequence to a machine learning method  \citep{whitehouse2024tree} may improve parameter estimates.

% ancient samples
Our approach relies on the locations of the current-day samples. While we have shown that we can learn much about the geographical history of a species with this approach, its accuracy is necessarily limited. For example, if historical parts of the range are undersampled in a particular dataset the method will struggle to locate ancestors in these regions, particularly further into the past, as we saw with the Iberian relicit samples with ancestry from a putative north African refugia. Similarly, if a species' range has shifted such that few present day individuals exist in portions of the historic range, as is likely the case of \textit{A. thaliana} in central Africa \citep{fulgione2018archaic}, we will often not infer ancestors to be in the currently sparsely-occupied portions. Other large scale movements, such as one ancestry replacing another (e.g., the non-relicts replacing the relicts in much of western Europe), may also partially obscure the geographic locations of ancestors. Over the past decade we have learned about numerous large-scale movements in humans alongside the expansions of archaeological cultures, a fact fairly hidden from view by contemporary samples that only ancient DNA could bring to light \citep{slatkin2016ancient,reich2018we}. One obvious way to improve our method then, is to include ancient samples. Given that it is now possible to include high-coverage ancient genomes in tree sequences \citep{speidel2021inferring,wohns2021unified}, it is straightforward to include this information in our likelihoods (ancient samples are treated as any other, we calculate the shared times of these lineages with themselves and with all other sample lineages), influencing both our dispersal estimates and inferred ancestral locations. This should help, in particular, in locating ancestors that are closely related to the ancient samples and for detecting large-scale movements, such as range expansions, contractions, and replacements.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Materials and Methods}
\label{sec:methods}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Here we describe our methods to estimate dispersal rates and locate genetic ancestors and how we applied these to simulations and \textit{Arabidopsis thaliana}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Dispersal rate}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{The probability distrubition of the sample locations given a tree and the dispersal rate}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Let $\mathbf{L}$ be the $n\times d$ matrix of sample locations, with $n$ samples in $d$ dimensions.
We first derive the probability distribution of $\mathbf{L}$ given a tree topology (a directed graph with $n-1$ nodes and $2n-2$ edges), $\mathcal{G}$, and associated branch lengths (a vector of $2n-2$ times), $\mathcal{T}$, which together describe the coalescent history of the sample, and the dispersal rate ($d\times d$ covariance matrix), $\mathbf{\Sigma}$. 
We derive this probability distribution, $\mathbb{P}(\mathbf{L} | \mathcal{G},\mathcal{T},\mathbf{\Sigma})$, by compounding the normally-distributed dispersal events each generation to give Brownian motion down the tree in a similar manner to phylogenetic least squares regression \citep{grafen1989phylogenetic,harmon2019phylogenetic}.

Let $\mathbf{S}_{\mathcal{G},\mathcal{T}}$ be the $n\times n$ matrix of shared  times (in generations) between each pair of sample lineages in the tree defined by $\mathcal{G}$ and $\mathcal{T}$. Then, assuming per generation dispersal is multivariate normal with mean displacement $\mathbf{0}$ and covariance matrix $\mathbf{\Sigma}$, the probability distribution of the locations depends only on the shared times in the tree, $\mathbb{P}(\mathbf{L} | \mathcal{G},\mathcal{T}, \mathbf{\Sigma}) = \mathbb{P}(\mathbf{L} | \mathbf{S}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma})$.
%
% To write this as a probability distribution, let $\bm{\ell}_i = [\ell_{i,1} \; \ell_{i,2} \; ... \;\ell_{i,d}]^\intercal$ be the $d$-dimensional location of sample $i$, so that $\mathbf{L} = [\bm{\ell}_1\; \bm{\ell}_2\; \dots \; \bm{\ell}_n]^\intercal$, and let $\bm{\ell} = [\ell_{1,1} \; \ell_{2,1} \; ... \; \ell_{n,1} \; \ell_{1,2} \; \ell_{2,2} \; ... \; \ell_{n,2} \; ... \; \ell_{1,d} \; \ell_{2,d} \; ... \; \ell_{n,d}]^\intercal$ be the matrix of locations represented as a vector of length $nd$.
% Further, the locations conditioned on the shared times and dispersal rate, $\mathbb{P}(\mathbf{L} | \mathbf{S}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma})$, are normally distributed
In fact, the locations are normally distributed
%
% \begin{equation}\label{eq:PS}
% % \begin{aligned}
% % \mathbb{P}(\mathbf{L} | \mathcal{G},\mathcal{T}, \mathbf{\Sigma}) = \mathbb{P}(\bm{\ell} | \mathbf{S}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma}) \\
% \mathbb{P}(\bm{\ell} | \mathbf{S}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma}) = \mathcal{N}\left(\mathbf{D} \widehat{\bm{\ell}_A}, \mathbf{S}_{\mathcal{G},\mathcal{T}} \otimes \mathbf{\Sigma} \right),
% % \end{aligned}
% \end{equation}
%
around the maximum likelihood estimate for the location of the most recent common ancestor,
% 
% \begin{equation}\label{eq:sa}
$\widehat{\bm{\ell}_A} = [(\mathbf{1}^\intercal \mathbf{S}_{\mathcal{G},\mathcal{T}}^{-1} \mathbf{1})^{-1}(\mathbf{1}^\intercal \mathbf{S}_{\mathcal{G},\mathcal{T}}^{-1} \mathbf{L})]$,
% \end{equation}
%
with a covariance that is the (Kronecker) product of the shared times and the dispersal rate, $\mathbf{S}_{\mathcal{G},\mathcal{T}} \otimes \mathbf{\Sigma}$.
Here $\mathbf{1}$ is a column vector of $n$ ones. %and $\mathbf{D}$ is a $nm\times m$ design matrix whose $i,j^\mathrm{th}$ entry is 1 if $(j-1) n < i \leq j n$ and 0 otherwise.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Mean centering}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

We can remove any dependence on the (unknown) location of the most recent common ancestor, $\bm{\ell}_A$, by mean centering the data \citep[e.g.,][]{lee2017distinguishing}.
In particular, we subtract the mean sample location from each sample location and drop one of the samples (as we lose a degree of freedom by centering).
The mean-centered locations can be computed with $\mathbf{X}=\mathbf{M}\mathbf{L}$, where $\mathbf{M}$ is an $(n-1)\times n$ matrix with $(n-1)/n$ on the diagonal and $-1/n$ elsewhere.
The mean-centered shared times are $\mathbf{V}_{\mathcal{G},\mathcal{T}}=\mathbf{M}\mathbf{S}_{\mathcal{G},\mathcal{T}}\mathbf{M}^\intercal$. %= n \mathbf{I} - \tfrac{1}{n}\mathbf{1}^\intercal \mathbf{1}$ with $\mathbf{I}$ an $n\times n$ identity matrix.
% We only use those (sub)trees (see below) with $n>1$ samples (i.e., trees with only one sample contain no information about the dispersal rate if we do not know where the ancestor was).

Similar to above, the probability distribution of the mean-centered locations only depends on the shared times, $\mathbb{P}(\mathbf{X} | \mathcal{G},\mathcal{T}, \mathbf{\Sigma}) = \mathbb{P}(\mathbf{X} | \mathbf{V}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma})$.
These mean-centered locations are normally distributed around $\mathbf{0}$ with covariance $\mathbf{V}_{\mathcal{G},\mathcal{T}} \otimes \mathbf{\Sigma}$.
%
% \begin{equation}\label{eq:PScentered}
%  \mathbf{x} | \mathbf{V}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma} \sim \mathcal{N}\left(\mathbf{0}, \mathbf{V}_{\mathcal{G},\mathcal{T}} \otimes \mathbf{\Sigma} \right),
% \end{equation}
% %
% where $\mathbf{x}$ is the vector representation of $\mathbf{X}$ (as $\bm{\ell}$ is to $\mathbf{L}$).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Chopping the tree}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

We will usually want to only consider lineages more recently than some cutoff time, $T$, since deeper genealogical history may contain little geographic information \citep{wilkins2004separation}.
When this is the case we cut the full tree of at $T$, leaving us with $n_T \in [1, n ] $ subtrees, where all coalescent times are $\leq T$.
Treating each subtree independently, the probability distribution of the sample locations is the product of the probability distributions for the sample locations in each subtree, 
$$
\mathbb{P}(\mathbf{X} | \mathcal{G}, \mathcal{T}, \mathbf{\Sigma}, T) = \prod_{i=1}^{n_T} \mathbb{P}(\mathbf{X}_i | \mathcal{G}_i, \mathcal{T}_i, \mathbf{\Sigma}) = \prod_{i=1}^{n_T} \mathbb{P}(\mathbf{X}_i | \mathbf{V}_{\mathcal{G}_i, \mathcal{T}_i}, \mathbf{\Sigma}),
$$
where $\mathcal{G}_i$ is the topology, $\mathcal{T}_i$ is the branch lengths, $\mathbf{V}_{\mathcal{G}_i, \mathcal{T}_i}$ is the (mean-centered) shared times, and $\mathbf{X}_i$ is the (mean-centered) sample locations for subtree $i$.
The (mean-centered) locations in subtree $i$ are normally-distributed
%
% \begin{equation}\label{eq:PScentered_chopped}
% \mathbf{x}_i | \mathbf{V}_{\mathcal{G}_i, \mathcal{T}_i}, \mathbf{\Sigma} \sim \mathcal{N}\left(\mathbf{0}, \mathbf{V}_{\mathcal{G}_i,\mathcal{T}_i} \otimes \mathbf{\Sigma} \right). %= \mathcal{N}\left(\mathbf{0}, \mathbf{V}_{\mathcal{G},\mathcal{T}}_T \otimes \mathbf{\Sigma} \right),
% \end{equation} 
% This has the added benefit of expressing the probability as a function of smaller submatrices of shared evolutionary times, which are faster to invert.
%where $\mathbf{V}_{\mathcal{G},\mathcal{T}}_T = \sum_{i=1}^{\eta} \mathbf{V}_{\mathcal{G},\mathcal{T}}_i^{-1}$.
around $\mathbf{0}$ with covariance $\mathbf{V}_{\mathcal{G}_i,\mathcal{T}_i} \otimes \mathbf{\Sigma}$.

Note that to mean center we need at least two samples in a subtree. Given that we do not know the root location, when there is only one sample we do not have enough information to infer a dispersal rate from this subtree. We therefore ignore subtrees with only one sample.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% 
\subsubsection*{Importance sampling}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

The calculations above, which give the likelihood of the dispersal rate given the sample locations, are all predicated on knowing the tree with certainty, which will not be the case when inferring trees from genetic data.
Further, inferring a tree may involve assumptions (such as panmixia) that are inconsistent with the model we are using.
To deal with uncertainty and sampling bias, we calculate the likelihood of our parameters given the data using importance sampling, a likelihood ratio, and Monte Carlo approximation.
Importance sampling corrects the expectation of the likelihood by reweighting draws from an `incorrect' (proposal) distribution to match the `correct' (target) distribution. 
These importance weights downweight draws from the proposal distribution that are likely under the proposal distribution but unlikely under the target distribution and upweight draws that are likely in the target distribution and unlikely under the proposal distribution. 
Importance sampling has been applied to genealogies in a number of population genetic settings as coalescent models provide convenient priors on trees and it is often challenging to sample genealogies consistent with data \citep{griffiths1997computational,stephens2000inference,meligkotsidou2007postprocessing}. 
See \cite{stern2019approximate,stern2020disentangling} for recent applications of these ideas to marginal trees inferred along a recombining sequence, whose general approach we follow below.  

The data we have are the locations of the samples, $\mathbf{L}$, and their haplotypes, $\mathbf{H}$ (a matrix with genomes as rows and sites as columns).
From this data we want to infer two unknowns: a tree topology, $\mathcal{G}$, and the associated branch lengths, $\mathcal{T}$.
We then want to use these unknowns to estimate the likelihood of the dispersal rate, $\mathbf{\Sigma}$.
Put another way, we want to estimate $\mathbb{E}_{\mathcal{G},\mathcal{T} | \mathbf{\Sigma}} \left[ \mathbb{P}(\mathbf{L},\mathbf{H} | \mathcal{G},\mathcal{T},\mathbf{\Sigma}) \right]$, which is the likelihood of our parameters given the data, integrated over the unknowns. 

Current methods to infer tree topologies and branch lengths along a recombining sequence \citep{rasmussen2014genome,speidel2019method,kelleher2019inferring,wohns2021unified} assume panmictic, well-mixed populations. 
This implies we cannot sample topologies and branch lengths, $\mathcal{G}$ and $\mathcal{T}$, under our spatial model, creating  bias.
We correct this bias with importance sampling, weighting each likelihood by the probability of the topology and branch lengths under our spatial model, $\mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{\Sigma})$, relative to their probability under the panmictic model we are sampling from, $\mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{H}, \mathrm{panmixia})$.
This allows us to integrate over topology and branch lengths from the panmictic model, giving a dispersal rate likelihood
%
\begin{equation}
\begin{aligned}
\mathrm{L}(\mathbf{\Sigma}) &:= \mathbb{E}_{\mathcal{G}, \mathcal{T} | \mathbf{\Sigma}} \left[ \mathbb{P}(\mathbf{L},\mathbf{H} | \mathcal{G},\mathcal{T},\mathbf{\Sigma}) \right] \\
&= \mathbb{E}_{ \mathcal{G}, \mathcal{T} | \mathbf{H}, \mathrm{panmixia}} \left[ \frac{\mathbb{P}(\mathbf{L},\mathbf{H} | \mathcal{G}, \mathcal{T},\mathbf{\Sigma}) \mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{\Sigma})}{\mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{H}, \mathrm{panmixia})} \right].
\end{aligned}
\end{equation}

The probabilities containing the genetic data, $\mathbf{H}$, are complicated to calculate. 
To simplify we divide by the likelihood of panmixia given the data, $\mathrm{L}(\mathrm{panmixia}) = \mathbb{P}(\mathbf{H}|\mathrm{panmixia})$, to work with the likelihood ratio
%
\begin{equation}
\begin{aligned}
\mathrm{LR}(\mathbf{\Sigma}) &:= \frac{\mathrm{L}(\mathbf{\Sigma})}{\mathrm{L}(\mathrm{panmixia})}\\
&= \mathbb{E}_{\mathcal{G}, \mathcal{T} | \mathbf{H}, \mathrm{panmixia}} \left[ \frac{\mathbb{P}(\mathbf{L}, \mathbf{H} | \mathcal{G}, \mathcal{T},\mathbf{\Sigma}) \mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{\Sigma})}{\mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{H}, \mathrm{panmixia}) \mathbb{P}(\mathbf{H} | \mathrm{panmixia})} \right] \\
&= \mathbb{E}_{\mathcal{G}, \mathcal{T} | \mathbf{H}, \mathrm{panmixia}} \left[ \frac{\mathbb{P}(\mathbf{L},\mathbf{H} | \mathcal{G}, \mathcal{T},\mathbf{\Sigma}) \mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{\Sigma})}{\mathbb{P}(\mathbf{H}, \mathcal{G}, \mathcal{T} | \mathrm{panmixia})} \right] \\
&= \mathbb{E}_{\mathcal{G}, \mathcal{T} | \mathbf{H}, \mathrm{panmixia}} \left[ \frac{\mathbb{P}(\mathbf{H} | \mathcal{G}, \mathcal{T}) \mathbb{P}(\mathbf{L} | \mathcal{G}, \mathcal{T},\mathbf{\Sigma}) \mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{\Sigma})}{\mathbb{P}(\mathbf{H} | \mathcal{G},\mathcal{T})\mathbb{P}(\mathcal{G}, \mathcal{T} | \mathrm{panmixia})} \right] \\
&= \mathbb{E}_{\mathcal{G}, \mathcal{T} | \mathbf{H}, \mathrm{panmixia}} \left[ \frac{ \mathbb{P}(\mathbf{L} | \mathcal{G}, \mathcal{T}, \mathbf{\Sigma}) \mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{\Sigma})}{\mathbb{P}(\mathcal{G}, \mathcal{T} | \mathrm{panmixia})} \right].
%&= \mathbb{E}_{T|\mathbf{H},G,\mathrm{panmixia}} \left[ \frac{ \mathbb{P}(\mathbf{L}, T | G, P)}{\mathbb{P}(\mathcal{T} | \mathcal{G},\mathrm{panmixia})} \right],
\end{aligned}
\end{equation}
%
The third step assumes the genetic data ($\mathbf{H}$) is conditionally independent of the spatial parameters ($\mathbf{\Sigma}$ or panmixia) and locations ($\mathbf{L}$) given the tree ($\mathcal{G}$ and $\mathcal{T}$).
We can approximate this expectation using Monte Carlo sampling,
%
\begin{equation}\label{eq:LRmontecarlo1}
\widehat{\mathrm{LR}(\mathbf{\Sigma})} = \frac{1}{M} \sum_{m=1}^{M} \frac{ \mathbb{P}(\mathbf{L} | \mathcal{G}_m, \mathcal{T}_m, \mathbf{\Sigma}) \mathbb{P}(\mathcal{G}_m, \mathcal{T}_m | \mathbf{\Sigma})}{\mathbb{P}(\mathcal{G}_m, \mathcal{T}_m | \mathrm{panmixia})},
\end{equation}
%
where the genealogies and branch lengths, $\mathcal{G}_m$ and $\mathcal{T}_m$, are sampled from the panmictic model, $\mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{H}, \mathrm{panmixia})$, using Markov chain Monte Carlo.

We make two final simplifications.
First, we will use a model of branching Brownian motion for $\mathbb{P}(\mathcal{G}, \mathcal{T} | \mathbf{\Sigma})$ and the standard neutral coalescent for $\mathbb{P}(\mathcal{G}, \mathcal{T} | \mathrm{panmixia})$.
Under both of these models the probability of the topology is equivalent (and uniform), $\mathbb{P}(\mathcal{G} | \mathbf{\Sigma}) = \mathbb{P}(\mathcal{G} | \mathrm{panmixia})$.
Second, we will use \texttt{Relate} \citep{speidel2019method} to infer topologies and branch lengths.
\texttt{Relate} returns a single topology and allows resampling over branch lengths conditional on this topology.
We therefore take the topology as given and integrate only over branch lengths.
Putting these two simplifications together, the likelihood ratio becomes
%
\begin{equation}\label{eq:LRmontecarlo}
\begin{aligned}
\widehat{\mathrm{LR}(\mathbf{\Sigma})} &= \frac{1}{M} \sum_{m=1}^{M} \frac{ \mathbb{P}(\mathbf{L} | \mathcal{G}_m, \mathcal{T}_m, \mathbf{\Sigma}) \mathbb{P}(\mathcal{G}_m, \mathcal{T}_m | \mathbf{\Sigma})}{\mathbb{P}(\mathcal{G}_m, \mathcal{T}_m | \mathrm{panmixia})}\\
&= \frac{1}{M} \sum_{m=1}^{M} \frac{ \mathbb{P}(\mathbf{L} | \mathcal{G}_m, \mathcal{T}_m, \mathbf{\Sigma}) \mathbb{P}(\mathcal{T}_m | \mathcal{G}_m, \mathbf{\Sigma}) \mathbb{P}(\mathcal{G}_m | \mathbf{\Sigma})}{\mathbb{P}(\mathcal{T}_m | \mathcal{G}_m, \mathrm{panmixia}) \mathbb{P}(\mathcal{G}_m | \mathrm{panmixia})}\\
&= \frac{1}{M} \sum_{m=1}^{M} \frac{ \mathbb{P}(\mathbf{L} | \mathcal{G}_m, \mathcal{T}_m, \mathbf{\Sigma}) \mathbb{P}(\mathcal{T}_m | \mathcal{G}_m, \mathbf{\Sigma})}{\mathbb{P}(\mathcal{T}_m | \mathcal{G}_m, \mathrm{panmixia}) }\\
&\approx \frac{1}{M} \sum_{m=1}^{M} \frac{ \mathbb{P}(\mathbf{L} | \mathcal{G}, \mathcal{T}_m, \mathbf{\Sigma}) \mathbb{P}(\mathcal{T}_m | \mathcal{G}, \mathbf{\Sigma})}{\mathbb{P}(\mathcal{T}_m | \mathcal{G}, \mathrm{panmixia})},
\end{aligned}
\end{equation}
%
where the branch lengths, $\mathcal{T}_m$, are sampled from $\mathbb{P}(\mathcal{T} | \mathbf{H}, \mathrm{panmixia}, \mathcal{G})$ using Markov chain Monte Carlo.
Note that because the probability of a topology is the same in the two models, and therefore cancels out, our method can immediately be extended to integrate over both topologies and branch lengths.
Future work could integrate over both topologies and branch lengths with other software \citep{rasmussen2014genome, wohns2021unified} or by running \texttt{Relate} on subsamples of the genomes and integrating over subsamples. 

% To find the maximum likelihood estimate of $\mathbf{\Sigma}$ we numerically search for the $\mathbf{\Sigma}$ that minimizes the negative log likelihood ratio. 
% % We measure the variance around this estimate with the Hessian matrix of the likelihood surface returned by the numerical search. 

We next show how we derive the probability distributions of the branch lengths in the tree that we use as the weights in the importance sampler, $\mathbb{P}(\mathcal{T} | \mathcal{G}, \mathrm{panmixia})$ and $\mathbb{P}(\mathcal{T} | \mathcal{G}, \mathbf{\Sigma})$.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{The probability distribution of the branch lengths given a tree topology and panmixia}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\texttt{Relate} and other tree construction methods infer the branch lengths in the trees under a model of panmixia with variable population size through time (with this demographic history inferred from the same haplotypes as part of the method). 
The probability distribution of these branch lengths given the tree topology and panmixia, $\mathbb{P}(\mathcal{T} | \mathcal{G}, \mathrm{panmixia})$, can be derived from the standard neutral coalescent allowing for (stepwise) changing effective population size \citep{griffiths1994sampling,meligkotsidou2007postprocessing}.
 
Let the number of samples be $n$ and let $u_i$ be the $i^\mathrm{th}$ most recent coalescence time (as determined by the branch lengths).
The coalescence times are produced by a Markov process, where the probability of the $i^{th}$ coalescence being at time $u_i$ is independent of the other coalescence times once conditioned on the time of the previous coalescence, $u_{i-1}$.
The probability distribution of the branch lengths can therefore be written
%
\begin{equation}
\mathbb{P}(\mathcal{T} | \mathcal{G},\mathrm{panmixia}) = \prod_{i=1}^{n-1} \mathbb{P}(U_i = u_i | U_{i-1} = u_{i-1}).
\end{equation}
%
Between time $u_{i-1}$ and $u_i$ we have $n-(i-1)$ lineages.
The coalescence time $u_i$ is then approximately distributed as a time-inhomogeneous exponential random variable with instantaneous rate $\lambda(u) = \binom{n-(i-1)}{2}/[2N(u)]$, where $N(t)$ is the effective population size at time $t$ in the past, giving
%
\begin{equation}
\begin{aligned}
& \mathbb{P}(U_i = u_i | U_{i-1} = u_{i-1})\\
& = \binom{n-(i-1)}{2}\frac{1}{2N(u_i)}\exp\left[ -\binom{n-(i-1)}{2} [\Lambda(u_i) - \Lambda(u_{i-1})] \right]
\end{aligned}
\end{equation}
%
with
%
\begin{equation}
\Lambda(u) = \int_0^u \frac{1}{2N(t)} \mathrm{d}t
\end{equation}
the cumulative coalescence rate up to time $u$.
%
%The probability of no coalescence between time $u_{i-1}$ and $u$ is then $\exp[-\int_{u_{i-1}}^u \lambda(t) \mathrm{d}t] = \binom{n-(i-1)] [\Lambda(u) - \Lambda(u_{i-1})]/2$.

When the effective population size is piecewise constant, split into $K$ epochs with effective population size $N_k$ between time $\tau_k$ and $\tau_{k+1}$, then the cumulative coalescence rate is \citep[see Appendix A of][]{stern2020disentangling}
%
\begin{equation}
\Lambda(u) = \left[ \sum_{k=0}^{b(u)} \frac{1}{2N_k} (\tau_{k+1} - \tau_k) \right] + \frac{1}{2N_{b(u)+1}} (u - \tau_{b(u)+1}),
\end{equation}
%
where $b(u) = \mathrm{max}(k \in \{0,1,2,...,K-1\} : u > \tau_{k+1} )$ is the last epoch to end before time $u$.

If we only want the probability density of coalescence times back as far as time $T$, we consider only these $m \leq n-1$ coalescence times and then multiply the probability density of these times by the probability of no coalescence between time $u_m$ and $T$,
%
\begin{equation}
\begin{aligned}
& \mathbb{P}(\mathcal{T} | \mathcal{G},\mathrm{panmixia},T)\\
& = \exp\left[ -\binom{n-m}{2} [\Lambda(T) - \Lambda(u_m)] \right] \prod_{i=1}^{m} \mathbb{P}(U_i = u_i | U_{i-1} = u_{i-1}).
\end{aligned}
\end{equation}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{The probability distribution of the branch lengths given a tree topology and our spatial model}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

A number of spatial models of the coalescent have been developed. 
However, for our purposes many are computational challenging to work with and so we pursue a simpler, analytically-tractable approximation called branching Brownian motion (BBM), or the Brownian-Yule process. 
Such approximations have been used previously to approximate genealogies on short timescales \citep{edwards1970estimation,rannala1996probability,meligkotsidou2007postprocessing}. 
We view the rate of branching ($\lambda$) in the BBM process as a nuisance parameter, but informally it is proportional to the inverse of the local population density and so it might be a useful parameter to explore in future work \citep[for a recent application see][]{ringbauer2017inferring}.

Let $T$ be the time in the past we wish to start the process (which in practice will be no greater than the time to the most recent common ancestor in the tree).
Let $n_0$ be the number of ancestral lineages at this time.
Taking a forward in time perspective, let $u'_i$ be the time at which we go from $i$ to $i+1$ lineages ($u_i' = T - u_{n - i}$).
Then under a pure birth (Yule) process with per capita birth rate $\lambda$ the joint probability density of the $n-n_0$ birth times and ending up with $n$ samples (i.e., no birth once $n$ lineages existed) is 
%
\begin{equation}\label{eq:fun}
\begin{aligned}
  \mathbb{P}(\mathbf{u}', n | \lambda, n_0, T) &=  \left( \prod_{i=n_0}^{n-1} \lambda i \exp[-\lambda i (u_i' - u_{i-1}')] \right)   \exp[-\lambda n (T - u_{n-1}')] \\
& = \lambda^{n-n_0} \frac{(n-1)!}{(n_0-1)!} \exp \left[-\lambda \left( n T + \sum_{i=n_0}^{n-1}u'_i \right) \right],
\end{aligned}
\end{equation}
%
where we have defined the starting time of the process $u_{n_0-1}' = 0$. 

The probability that $n_0$ lineages produce exactly $n$ lineages in time $T$ under a pure birth process is given by a negative binomial with $n-n_0$ successes, $n_0$ failures, and success probability $1 - \exp(-\lambda T)$,
%
\begin{equation}\label{eq:pn}
\mathbb{P}(n|\lambda,n_0,T) = \binom{n-1}{n-n_0} \exp(-\lambda T n_0) [1-\exp(-\lambda T)]^{n-n_0}.
\end{equation}
%

The probability distribution of the branch lengths under the Yule process, which we take to be the probability of the branch lengths given the topology and the spatial model, is therefore proportional to the ratio of Equations \eqref{eq:fun} and \eqref{eq:pn},
%
\begin{equation}\label{eq:pTSigma}
\begin{aligned}
\mathbb{P}(T | G, \mathbf{\Sigma}) &= \mathbb{P}(\mathbf{u'} | \lambda,n_0,T,n) \\
& \propto \frac{\mathbb{P}(\mathbf{u}', n | \lambda, n_0, T)}{\mathbb{P}(n|\lambda,n_0,T)} \\
& = (n-n_0)! \left(\frac{\lambda \exp(-\lambda T)}{1-\exp(-\lambda T)}\right)^{n-n_0} \exp\left(-\lambda \sum_{i=n_0}^{n-1}u'_i \right).
\end{aligned}
\end{equation}
%%
%When $\mathcal{T}$ is the time to the most recent common ancestor (so $n_0=2$) the ratio in \eqref{eq:pTSigma} reduces to 
%%
%\begin{equation}
%(n-2)! \left(\frac{\lambda \exp(-\lambda T)}{1-\exp(-\lambda T)}\right)^{n-2} \exp\left(-\lambda \sum_{i=2}^{n-1}u_i \right).
%\end{equation}
%%
Note that the branching times are independent of the dispersal rate given the topology, $\mathbb{P}(T | G, \mathbf{\Sigma})=\mathbb{P}(T | G)$, under this relatively simple spatial model.
See \cite{edwards1970estimation,rannala1996probability,meligkotsidou2007postprocessing} for more on the Yule process.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Conditioning on sampling locations}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

As many studies choose sampling locations before choosing samples, we would like to condition on sampling locations in our inferences of dispersal rate, i.e., use the conditional likelihood $\mathbb{P}(\mathbf{L} | \mathcal{G},\mathcal{T}, \mathbf{\Sigma})/\mathbb{P}(\mathbf{L} | \mathbf{\Sigma})$.
The denominator, $\mathbb{P}(\mathbf{L} | \mathbf{\Sigma})$, is the probability distribution of the locations of the descendent tips of a branching Brownian motion given the dispersal rate.
This is the numerator, $\mathbb{P}(\mathbf{L} | \mathcal{G},\mathcal{T}, \mathbf{\Sigma})$, after the genealogy and branch times have been integrated out. 
Calculating the integrals over genealogies and branch times could be achieved by Monte Carlo simulation, but would be computationally intensive. 
However, as shown by \citep{meligkotsidou2007postprocessing}, in one dimension (say $x$) the probability distribution of the locations, $\mathbb{P}(\mathbf{L} | \sigma_x)$, is independent of dispersal, $\sigma_x$, if a scale invariant prior is placed on the branching rate. 
Intuitively, this follows from the fact that if we do not {\it a priori} know the scale of the coalescence rate then we do not know {\it a priori} how long the branch lengths are, and so we do not know how fast lineages need to disperse to get to where they are today. 
Therefore, in that case, conditioning on the locations does not change the likelihood surface of $\sigma_x$ and so can be ignored in the inference of dispersal. 
This result also holds in two dimensions if we constrain $\sigma_x=\sigma_y$. 
Sadly, it does not hold in general for an arbitrary dispersal matrix $\mathbf{\Sigma}$. 
For example, if our samples (the end points of a BBM) were distributed widely along the $x$ axis but varied little in displacement along the $y$ axis, this would be more likely if $\sigma_y/\sigma_x \ll 1$. 
Thus, while the overall magnitude of dispersal is not affected by conditioning on the sampling locations, the anisotropy of dispersal may be.  

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%% 
\subsubsection*{Composite likelihood over loci}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

To use information from multiple loci, e.g., to estimate a genome-wide dispersal rate, we can multiply the likelihood ratios together to give a composite likelihood ratio.
This is not the full likelihood because it ignores correlations in the genealogical and spatial processes between loci \citep{hudson2001two,larribe2011composite,varin2011overview}. 
The maximum likelihood of genome-wide parameters from composite likelihood are known to be statistically consistent in the presence of such correlations \citep{wiuf2006consistency}, but the likelihood surface is overly peaked due to ignoring the correlations in the data. 
However, in practice we try to take loci that are far enough apart that their trees are only weakly correlated more than a few tens of generations back.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Maximum likelihood estimates}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% \textcolor{red}{mo: reworked this to give MLEs over subtrees and loci, which helps tie into new section below}

The likelihood of the dispersal rate given the shared times and sample locations in a tree is $\mathbb{P}(\mathbf{L} | \mathbf{S}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma})$.
This implies the maximum likelihood estimate of the dispersal rate is
%
\begin{equation}\label{eq:MLEsigma}
\widehat{\mathbf{\Sigma}} = \frac{(\mathbf{L} - \mathbf{1}\widehat{\bm{\ell}_A})^\intercal \mathbf{S}_{\mathcal{G},\mathcal{T}}^{-1}(\mathbf{L} - \mathbf{1}\widehat{\bm{\ell}_A})}{n}.
\end{equation}
%
This estimate is, however, biased because it relies on the unknown location of the root of the tree. 
Using the mean-centered likelihood, $\mathbb{P}(\mathbf{X} | \mathbf{V}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma})$, the restricted maximum likelihood estimate of the dispersal rate given for subtree $j$ is
%
\begin{equation}\label{eq:RMLEsigma}
\widehat{\mathbf{\Sigma}}^* = \frac{\mathbf{X}^\intercal \mathbf{V}_{\mathcal{G},\mathcal{T}}^{-1}\mathbf{X}}{n-1},
\end{equation}
%
which is unbiased and equivalent to $n\widehat{\mathbf{\Sigma}}/(n-1)$.

When we chop a tree into subtrees and multiply the (mean-centered) likelihoods over subtrees, the (restricted) maximum likelihood dispersal rate becomes a weighted average of the (restricted) maximum likelihood dispersal rates from each subtree,
\begin{equation}\label{eq:RMLEsigma_subtrees}
  \widehat{\mathbf{\Sigma}}^* = \frac{\sum_j(n_j-1)\widehat{\mathbf{\Sigma}}^*_j}{\sum_j (n_j - 1)},
  \end{equation}
where $\widehat{\mathbf{\Sigma}}^*_j$ is the (restricted) maximum likelihood dispersal rate from subtree $j$ (calculated as above) and $n_j$ is the number of samples in subtree $j$.

Similarly, when we then also multiply (mean-centered) likelihoods across loci the (restricted) maximum likelihood dispersal rate is a weighted average of the (restricted) estimates over the subtrees at all loci,
\begin{equation}\label{eq:RMLEsigma_subtrees_loci}
  \widehat{\mathbf{\Sigma}}^* = \frac{\sum_i\sum_j(n_{ij}-1)\widehat{\mathbf{\Sigma}}^*_{ij}}{\sum_i\sum_j (n_{ij} - 1)},
  \end{equation}
where $\widehat{\mathbf{\Sigma}}^*_{ij}$ is the restricted maximum likelihood dispersal rate from subtree $j$ at locus $i$ and $n_{ij}$ is the number of samples in subtree $j$ at locus $i$.

Unfortunately there is no analytical expression for the maximum likelihood dispersal rate with importance sampling, and so we instead numerically search for the dispersal rate (and branching rate, $\lambda$) which minimizes the negative log composite likelihood ratio (the negative log of the product of Equation \eqref{eq:LRmontecarlo} over loci).
We initialize the search with the maximum likelihood dispersal rate using just the first tree at each locus (Equation \ref{eq:RMLEsigma_subtrees_loci}).

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Best linear unbiased predictor}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

There is an analytical approximation to the maximum likelihood dispersal rate when we sample multiple trees per locus but ignore the importance weights (since we do not know the branching rate, $\lambda$).
Let $\widehat{\mathbf{\Sigma}}^*_{ij}$ be the (restricted) maximum likelihood dispersal rate from tree $j$ at locus $i$ (as calculated from Equation \ref{eq:RMLEsigma_subtrees}).
Averaging this over all trees and loci then gives a best linear unbiased predictor (BLUP) of the (restricted) maximum likelihood dispersal rate.
In the main text we numerically search for the (restricted) maximum likelihood dispersal rate, which allows us to include importance weights and more accurately infers simulated dispersal rates, but we implement the BLUP method in our software as it can be a useful way to quickly get a rough estimate of dispersal while still accounting for tree uncertainty.
% roughly twice the mle dispersal rate (sd) for thaliana with 1e4 cutoff

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Ancestor locations}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{The probability distribution of an ancestor locations given a tree and the dispersal rate}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

A point on a tree represents a genetic ancestor.
Choosing one such point, if the ancestor's lineage shares $\mathbf{s}_a$ time with each of the sample lineages and $s_a$ time with itself, then, by the properties of the conditional normal distribution, the probability that the ancestor was at location $\bm{\ell}_a$ is normally distributed
% %
% \begin{equation}\label{eq:Psa}
% \mathbb{P}(\bm{\ell}_a | \bm{\ell}, \mathbf{S}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma}, \mathbf{s}_a) \sim \mathcal{N}\left(\widehat{\bm{\ell}_a}, \widehat{\mathbf{S}} \right),
% \end{equation}
% %
% where
around
%
\begin{equation}\label{eq:sahat}
\widehat{\bm{\ell}_a} = \widehat{\bm{\ell}_A} + \mathbf{s}_a^\intercal \mathbf{S}_{\mathcal{G},\mathcal{T}}^{-1} (\mathbf{L} - \mathbf{1}\widehat{\bm{\ell}_A})
\end{equation}
%
% is the maximum likelihood estimate of the ancestor's location and
with covariance
%
\begin{equation}\label{eq:Chat}
\widehat{\mathbf{S}} = (s_a - \mathbf{s}_a^\intercal \mathbf{S}_{\mathcal{G},\mathcal{T}}^{-1} \mathbf{s}_a)\mathbf{\Sigma}.
\end{equation}
%
% is the uncertainty (covariance) in the maximum likelihood estimate.
% Here $\mathbf{S}_{\mathcal{G},\mathcal{T}}^{-1}$ is the generalized inverse of $\mathbf{S}_{\mathcal{G},\mathcal{T}}$.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Mean centering}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

This approach gives the correct mean (Equation \eqref{eq:sahat}) but the uncertainty (Equation \eqref{eq:Chat}) is incorrect because the approach implicitly assumes we know with certainty where the most recent common ancestor was.
%For instance, when there is only a single sample, Equation \eqref{eq:Chat} says that the uncertainty of the ancestor's location is $s_a(1 - 2s_a/T) \mathbf{\Sigma}$, which is maximized at $s_a = T/2$, when we should instead have an uncertainty of $(T-s_a)\mathbf{\Sigma}$ (i.e., the variance increases linearly as we go back in time).

The correct uncertainty is derived by mean centering.
The mean-centered sample locations, $\mathbf{X}$, and shared times among the samples, $\mathbf{V}_{\mathcal{G},\mathcal{T}}$, are described above.
The mean-centered vector of shared times between the ancestor and the samples is found by calculating the covariances between their mean-centered locations, giving $\mathbf{v}_a = \mathbf{M}(\mathbf{s}_a - \overline{\mathbf{s}}_i)$, where $M$ is the mean-centering matrix (defined above) and $\overline{\mathbf{s}}_i$ is the mean shared time of each sample over all others.
The mean-centered shared time of the ancestor with itself is found by calculating the variance in its mean-centered location, giving $v_a = s_a - 2\overline{\mathbf{s}}_a + \overline{\mathbf{s}}$, where $\overline{\mathbf{s}}_a$ is the mean shared time of the ancestor over all samples and $\overline{\mathbf{s}}$ is the average over all elements of $\mathbf{S}_{\mathcal{G},\mathcal{T}}$.

The probability the ancestor was at location $\bm{\ell}_a$ is then normally distributed with
%
% \begin{equation}\label{eq:PsaCentered}
% \mathbb{P}(\bm{\ell}_a | \mathbf{x}, \mathbf{V}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma}, \mathbf{v}_a, v_a) \sim \mathcal{N}\left(\widehat{\bm{\ell}_a}, \widehat{\mathbf{S}}^* \right),
% \end{equation}
% %
% where the maximum likelihood estimate has not changed but can be rewritten
mean as above but now written
\begin{equation}\label{eq:sahatCentered}
\widehat{\bm{\ell}_a} = \overline{\mathbf{L}} + \mathbf{v}_a^\intercal \mathbf{V}_{\mathcal{G},\mathcal{T}}^{-1} \mathbf{X} 
\end{equation}
%
% with $\bar{\mathbf{L}}$ the mean location of the samples. 
% The uncertainty in this estimate is
and covariance
%
\begin{equation}\label{eq:ChatCentered}
\widehat{\mathbf{S}}^* = \left( v_a - \mathbf{v}_a^\intercal \mathbf{V}_{\mathcal{G},\mathcal{T}}^{-1} \mathbf{v}_a \right) \mathbf{\Sigma},
\end{equation}
where $\overline{\mathbf{L}}$ is the mean location of the samples. 
%
This covariance gives the correct uncertainty.
For example, it produces a linear increase in the variance of the location as we move up from a sample when there is only a single sample, i.e., this is simple Brownian motion.
When there are two samples with a recent common ancestor at time $t$, the variance in ancestor location along this tree back to the most recent common ancestor is maximized at the most recent common ancestor, $t/2$, i.e., this is a Brownian bridge.
And so on.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Chopping the tree}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

As with the dispersal likelihood, to reduce dependencies on distant times we can chop the tree at some time $T$ and use only the subtree containing the ancestor of interest. 
Often the most recent common ancestor of the resulting subtree occurs more recently than $T$ but we want to infer the location all the way back to $T$, and so we need to infer the location of the common ancestral lineage past the most recent common ancestor. 
Fortunately, this method extends naturally to times beyond the most recent common ancestor, implicitly modelling simple Brownian motion (a fixed mean and linearly increasing variance) up the stem of the tree.
Similarly, the ancestor need not have descendants in the sample -- our method implicitly adds simple Brownian motion down the branch leading to a `hanging' ancestor (which is useful when we want to ignore the location of a sample and locate it or its ancestors using the remaining sample locations and the trees).
When there is only a single sample in a subtree we cannot mean center and instead model simple Brownian motion up from the sample.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Importance sampling}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Finally, to integrate over tree uncertainty and reduce bias from the way in which we sample trees we importance sample.
We do this by replacing the dispersal likelihood, $\mathbb{P}(\mathbf{L} | \mathcal{G},\mathcal{T}, \mathbf{\Sigma})$, in Equation \eqref{eq:LRmontecarlo} with the probability distribution for the ancestor's location, $\mathbb{P}(\bm{\ell}_a | \mathbf{X}, \mathbf{V}_{\mathcal{G},\mathcal{T}}, \mathbf{\Sigma}, \mathbf{v}_a, v_a)$,
%
\begin{equation}\label{eq:LRmontecarloAncestors}
\widehat{\mathrm{LR}(\bm{\ell}_a)} = \frac{1}{M} \sum_{m=1}^{M} \frac{ \mathbb{P}(\bm{\ell}_a | \mathbf{X}, \mathbf{V}_{\mathcal{G},\mathcal{T}_m}, \mathbf{\Sigma}, \mathbf{v}_a, v_a) \mathbb{P}(\mathcal{T}_m | \mathcal{G}, \mathbf{\Sigma})}{\mathbb{P}(\mathcal{T}_m | \mathcal{G},\mathrm{panmixia})}.
\end{equation}
%
We compute this using the maximum composite likelihood ratio estimates of $\mathbf{\Sigma}$ and $\lambda$.
% We measure the variance around this estimate with the Hessian matrix returned by the numerical search.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsubsection*{Best linear unbiased predictor}

As with dispersal, the above approach requires a numerical search for the most likely ancestor location because a sum of normal distributions (Equation \eqref{eq:LRmontecarloAncestors}) does not, in general, follow a tractable distribution.
An alternative approach, however, is to calculate the most likely ancestor location for each sampled tree (Equation \eqref{eq:sahatCentered}) and importance sample over these estimates.
Writing the most likely ancestor location given a sampled tree (Equation \ref{eq:sahatCentered}) as $\widehat{\bm{\ell}_a}(\mathcal{T}_m, \mathcal{G})$, we then have another estimate of the ancestor's location,
\begin{equation}\label{eq:LRmontecarloAncestorsBLUP}
\tilde{\bm{\ell}_a} = \frac{1}{M} \sum_{m=1}^{M} \frac{ \widehat{\bm{\ell}_a}(\mathcal{T}_m, \mathcal{G}) \mathbb{P}(\mathcal{T}_m | \mathcal{G}, \mathbf{\Sigma})}{\mathbb{P}(\mathcal{T}_m | \mathcal{G},\mathrm{panmixia})}.
\end{equation}
Equation \eqref{eq:LRmontecarloAncestorsBLUP} is a best linear unbiased predictor (BLUP) averaged over the importance weights. 
We can calculate this directly and it is therefore, in principle, faster to compute than the search over Equation \eqref{eq:LRmontecarloAncestors}. 
It is also independent of the dispersal rate since $\mathbb{P}(\mathcal{T}_m | \mathcal{G}, \mathbf{\Sigma})=\mathbb{P}(\mathcal{T}_m | \mathcal{G})$ under our spatial model.
In practice we find that the two estimators are very correlated, and the more correct numerical search over Equation \eqref{eq:LRmontecarloAncestors} is reasonably quick. 
Throughout the paper we use the numerical search over Equation \eqref{eq:LRmontecarloAncestors} but the BLUP method is also implemented in our software.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Simulations}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Our full pipeline is described in detail in a \texttt{Snakemake} \citep{molder2021sustainable} file at \url{https://github.com/mmosmond/spacetrees-ms}.

\subsubsection*{Spatially-explicit forward-time simulations}

We performed simulations in \texttt{SLiM} v4.0.1 \citep{haller2023slim} with tree sequence recording \citep{haller2019tree}.
The \texttt{SLiM} code was adapted from the \texttt{pyslim} spatial vignette (\url{https://tskit.dev/pyslim/docs/stable/vignette_space}) to model non-overlapping generations. 

Individuals are diploid for a single chromosome with $L=10^8$ basepairs and per basepair recombination rate $r=10^{-8}$.
Individuals exist in a two dimensional habitat, a square of width $W=100$ with reflecting boundaries.

Each generation begins with reproduction.
Each individual acts once as a `mother' and chooses a `father' at random (individuals are hermaphrodites), weighted by their mating weights.
The mating weight of an individual distance $d$ from a mother follows a 2-dimensional normal distribution centered on the mother with variance $\sigma_m^2=1/4$ in both directions and no covariance.
Individuals further apart than $3\sigma_m$ are ignored for efficiency.

Local density-dependence is modelled through competitive effects on fecundity.
The strength of competitive interaction between two individuals distance $d$ distance apart follows a 2-dimensional normal distribution centered on one of the individuals with variance $\sigma_c^2=1$ in both directions and no covariance.
We again ignore individuals more distant than $3\sigma_c$.
The number of offspring produced by a mother is Poisson, with mean $R/(1 + C/K)$, where $C$ is the sum of interaction strengths the mother experiences, $R=2$ is the mean number of offspring in the absence of competition, and $K=1$ is the local carrying capacity. 
Each offspring disperses from its mother by a random 2-dimensional normal deviate with variance $\sigma_d^2$ in each dimension and no covariance. 

Note that, given a lineage has an equal chance of descending from its mother or its father each generation, the true simulated dispersal rate is $\sigma^2 = \sigma_d^2/2 + (\sigma_d^2 + \sigma_m^2)/2 = \sigma_d^2 + \sigma_m^2/2$ \citep{smith2023dispersal} in each dimension.

After reproduction all adults die and all offspring become adults in the next generation.
We begin the population with $N_0=W^2 K$ individuals distributed uniformly at random across space.
We end the simulation, and output the tree sequence, after $4N_0$ generations.

% To analyze our ability to locate ancestors we remembered the locations of ancestors using the `initializeTreeSeq(retainCoalescentOnly=F)' and\\ %custom linebreak
% `treeSeqRememberIndividuals(permanent=F)' options.
% We then simplifying the tree sequence as above but with `keep\_unary=True', to keep all nodes (and their locations) that are ancestral to the sample.

\subsubsection*{True tree sequence of the sample}

Using \texttt{tskit} v0.5.6 \citep{kelleher2018efficient}, we load the tree sequence into \texttt{Python} v3.11.5, sample $k/2=50$ present-day diploid individuals ($k=100$ genomes) at random, and simplify the tree sequence to lineages ancestral to the sample.
We then use \texttt{pyslim} v0.6 (\url{https://github.com/tskit-dev/pyslim})to `recapitate' the tree sequence under the coalescent with recombination \citep{hudson2002generating} with effective population size $N_0$, ensuring all sampled lineages have coalesced.
This tree sequence represents the true genealogical history of the sample.

\subsubsection*{Inferred tree sequence of the sample}

Because we will not know the true tree sequence of any natural sample, we next use \texttt{msprime} v1.3.1 \citep{kelleher2016efficient} to layer neutral mutations on the tree sequence with per basepair per generation mutation rate $U=10^{-8}$.
We then write the genotypic data out as a VCF and use \texttt{Relate} v1.2.1 \citep{speidel2019method} to infer the tree sequence.
We give \texttt{Relate} the true uniform recombination map and mutation rate and an initial effective population size of $\theta/(4U)$, where $\theta$ is the observed mean genetic diversity calculated from the tree sequence with \texttt{tskit} \citep{ralph2020efficiently}.
We then feed the resulting output to \texttt{Relate}'s `EstimatePopulationSize' function to iteratively estimate a piece-wise constant effective population size and branch lengths, using 5 MCMC iterations (the default) and dropping 50\% of the trees (the default).
% We then use \texttt{Relate} to convert the anc/mut format to a \texttt{tskit} tree sequence for downstream processing. 

\subsubsection*{Processing the tree sequence}
  
At every 100th tree we use \texttt{Relate}'s `SampleBranchLengths' function to sample branch lengths from the posterior $M=1000$ times, using the true mutation rate and saving the output in Newick format.
We next load the Newick trees into \texttt{Python} as 1-tree \texttt{tskit} tree sequences with \texttt{tsconvert} (\url{https://github.com/tskit-dev/tsconvert}) and calculate the shared times and coalescent times.
We then chop these times at a cutoff $T$ and calculate various quantities that are used for inference: the shared time matrix of each subtree, the inverted mean-centered shared time matrix of each subtree, the log of the determinant of the mean-centered shared time matrix of each subtree, the coalescence times, and the log probability of the coalescence times.
This pre-processing speeds up the numerical search for maximum likelihood parameters.

\subsubsection*{Dispersal estimates}

We approximate the maximum composite likelihood estimate of the dispersal rate by numerically searching for the dispersal rate (and branching rate, $\lambda$) that minmizes the negative log of the product of likelihood ratios (Equation \eqref{eq:LRmontecarlo}) over sampled loci.
We use the L-BFGS-B method of \texttt{SciPy} v1.11.2 \citep{2020SciPy-NMeth} to numerically find the maximum, which allows us to impose bounds on the parameters.
We search for the maximum in terms of the standard deviation of dispersal in $x$ and $y$ (with a lower bound of $10^{-6}$ to prevent negative values) and the correlation between these two axes (with bounds of -0.99 and 0.99 to prevent estimates with absolute value greater than 1). 
We also place a lower bound of $10^{-6}$ on $\lambda$ to prevent negative branching rates.

\subsubsection*{Locating genetic ancestors}

To locate a genetic ancestor at a particular locus and time we numerically search for the location that maximizes the likelihood ratio (Equation \eqref{eq:LRmontecarloAncestors}), following the same approach as above but with the SLSQP method without bounds.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{\textit{Arabidopsis thaliana} data}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

Our full pipeline is described in detail in a \texttt{Snakemake} \citep{molder2021sustainable} file at \url{https://github.com/mmosmond/spacetrees-ms}.

\subsubsection*{Inferring the tree sequence}

We first downloaded 63 fastq sequences from \url{https://www.ebi.ac.uk/ena/browser/view/PRJEB55632} and \url{https://www.ebi.ac.uk/ena/browser/view/PRJEB55353} \citep{wlodzimierz2023cycles} and 3 genome assemblies: Col-0 and Ey15-2 from \url{https://www.ncbi.nlm.nih.gov/bioproject/PRJEB50694} and  Kew-1 from \url{https://www.ncbi.nlm.nih.gov/bioproject/PRJEB51511}.
We also downloaded a hard-masked TAIR10 reference genome from \url{https://zenodo.org/records/7326462}.

To extract the sample haplotypes from this data we roughly followed the methods of \cite{wlodzimierz2023cycles}.
We first used \texttt{hifiasm} v0.19.5 \citep{cheng2021haplotype,cheng2022haplotype} to assemble each of the 63 fastqs into contigs and convert the outputs to fasta.
We then used \texttt{ragtag} v2.1.0 \citep{alonge2022automated}, with the \texttt{minimap2} v2.26 \citep{li2018minimap2,li2021new} aligner and the TAIR10 reference, to scaffold each of the 63 assemblies.
We next extracted the 5 autosomes from each of the 66 assemblies, aligned these to the TAIR10 reference with \texttt{minimap2} and converted to sam/bam format with \texttt{samtools} v1.13 \citep{danecek2021twelve}.
We then used \texttt{bcftools} v1.13 \citep{danecek2021twelve} and the TAIR10 reference to create a VCF for each chromosome.
We also used \texttt{bcftools} to filter out sites with missing or heterozygote genotypes and filter to just biallelic single nucleodide polymorphisms (SNPs).
To get the haplotypes of each sample we used \texttt{bcftools} to convert the VCFs to hap/sample format and from this kept just one haplotype of each (now explicitly haploid) sample.

To polarize the SNPs we used a multi-species alignment for \textit{Arabidopsis thaliana}, \textit{Boechera stricta}, \textit{Arabidopsis lyrata}, and \textit{Malcomia maritima} (available at \url{https://doi.org/10.5281/zenodo.11456353}).
We used \texttt{bx-python} v0.8.9 (\url{https://github.com/bxlab/bx-python}) to create a fasta file containing the three outgroups from this alignment and, combined with the hap files described above, created the input files required by \texttt{est-sfs} \citep{keightley2018inferring}.
We then ran \texttt{est-sfs} v2.04, separately for each chromosome, using the Kimura 2-parameter model \citep{kimura1980simple} to get the probability that each major allele is ancestral.
This method includes information about polymorphism within the ingroup when assigning ancestral states.
Wherever the reference allele has a less than 1/2 probability of being ancestral we `flip' the alleles at that locus, so that 0 and 1 refer to ancestral and derived alleles, respectively, in our hap files.

To remove difficult to sequence regions of the genome we downloaded the hard-masked TAIR10 reference from \url{https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-56/fasta/arabidopsis_thaliana/dna/} and use \texttt{Relate} to mask the hap files.
We also downloaded a recombination map from \url{https://popsim-consortium.github.io/stdpopsim-docs/stable/catalog.html#sec_catalog_aratha_genetic_maps_salomeaveraged_tair10} \citep{10.7554/eLife.54967,salome2012recombination} and converted it into the form required by \texttt{Relate}.

We ran \texttt{Relate} v1.1.9 with the polarized, masked, haploid hap files and the recombination map to infer the genome-wide genealogies, with initial effective population size $10^5$ and mutation rate $7\times10^{-9}$ \citep{10.7554/eLife.54967}.
We then used \texttt{Relate} to infer a piece-wise constant effective population size and re-estimate branch lengths accordingly, using 10 MCMC iterations (the default is 5) and 50\% of the trees (default).
The final anc/mut output is converted into a \texttt{tskit} tree sequence with \texttt{Relate} for additional analyses.
We also use \texttt{Relate} to sample branch lengths 1000 times at every 100th tree on each chromosome (for a total of 878 trees).

To extract the information we need from the Newick files output by \texttt{Relate} we used \texttt{tsconvert} v0.1 (\url{https://github.com/tskit-dev/tsconvert}) to convert them into (one-tree) \texttt{tskit} tree sequences and \texttt{tskit} v0.5.2 to extract the shared times and coalescent times from each.
We then applied the time cutoffs, and for each subtree pre-computed the inverse and determinant of the mean-centered shared time matrix.
We also pre-calculated the probability of the coalescence times given the piecewise-constant effective population size inferred by \texttt{Relate}.
With these quantities in hand we estimate dispersal rates and locate genetic ancestors.

\subsubsection*{Dispersal estimates}

% We estimate the MLE of dispersal rate at each chosen tree as described above for simulations. 
% To reduce computation time we search for maximum composite likelihoods across trees for each chromosome separately, rather than genome-wide.
We estimated dispersal rates as described above for simulations.
We converted our estimates of dispersal rate from degrees squared to kilometres squared by multiplying the latitudinal estimates by $111^2$ and by multiplying the longitudinal estimates by $(111\cos(\bar{y} \pi/180))^2$, where $\bar{y}$ is the mean latitude across samples.

As one of our results is that rates of dispersal are higher along latitude than along longitude, we wanted to make sure that the lack of conditioning on sampling location did not drive this result. 
To do this we ran our method looking at only longitudinal dispersal, ignoring the sample latitudes, and then again looking at only latitudinal dispersal, ignoring the sample longitudes. 
This reduces the problem back to one dimension where the conditioning does not matter \citep[as discussed above and in][]{meligkotsidou2007postprocessing}. 
Doing this we find essentially identical dispersal rates as before, justifying our result of a larger latitudinal dispersal rate despite not conditioning on the sample locations. 

\subsubsection*{Locating genetic ancestors}

We locate ancestors as described above for simulations, but without imposing any bounds.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\subsection*{Data availability statement}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

All code used to perform the analyses in this study can be found at \url{https://github.com/mmosmond/spacetrees-ms}.
More information on how to run our method, \texttt{spacetrees}, is available at \url{https://github.com/osmond-lab/spacetrees}.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Acknowledgements}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

We thank Tyler Kent, Adrian Platts, and the Brassicales Map Alignment Project (DOE-JGI, http://bmap.jgi.doe.gov/) for the multispecies genome alignments,
Yan Wong, Ben Haller, and Peter Ralph for the \texttt{retainCoalescentOnly} and \texttt{permanent} updates to \texttt{SLiM}/\texttt{pyslim}, 
Leo Speidel for help with \texttt{Relate}, 
Vince Buffalo for an introduction to \texttt{Snakemake},
Fabrizio Menardo for pointing out an error in our lat/long to km conversion,
Puneeth Deraje and James Kitchens for helpful discussions,
and Nick Barton for healthy skepticism about per-locus estimates of ancestor locations at deep times.
%
Funding provided by Banting (Canada), the Center for Population Biology (UC Davis), the Natural Sciences and Engineering Research Council of Canada (RGPIN-2021-03207, awarded to MO), and the National Institute of General Medical Sciences of the National Institutes of Health (NIH R01 GM108779 and R35 GM136290, awarded to GC).
Computations were performed on the Niagara supercomputer at the SciNet HPC Consortium. 
SciNet is funded by: the Canada Foundation for Innovation; the Government of Ontario; Ontario Research Fund - Research Excellence; and the University of Toronto.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\section*{Competing interests}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%We have no financial and non-financial competing interests to declare.

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\bibliography{bib}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% \newpage
\section*{Supplementary Figures}
\label{sec:appendix}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% \newpage
%\section*{Supplementary Figures}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\setcounter{figure}{0}
\renewcommand{\thefigure}{S\arabic{figure}}
\newcounter{SIfig}
\renewcommand{\theSIfig}{S\arabic{SIfig}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%simulation dispersal estimates -- effect of importance sampling
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]
  %\captionsetup{labelformat=empty}
  \begin{center}
  \begin{tikzpicture}[remember picture]
  
    % 1000 importance samples

    % disp rates, no cutoff
    \node[] (Tnone) {
      \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone.pdf}    
    };
    % \node[left = 0cm of Tnone.north west, anchor = north west, yshift = -0.15cm] (A) {
    %   \textbf{A}
    % };
    \node[above = -0.75cm of Tnone, anchor=south] () {
      \footnotesize{$T = \infty$}
    };
  
    % disp rates, 1000 cutoff
    \node[right = -0.2cm of Tnone] (T1000) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000.pdf}
    };
    % \node[left = 0cm of T1000.north west, anchor = north west, yshift = -0.15cm] (B) {
    %   \textbf{B}
    % };
    \node[above = -0.75cm of T1000, anchor=south] () {
      \footnotesize{$T = 1000$}
    };
  
    % disp rates, 100 cutoff
    \node[right = -0.2cm of T1000] (T100) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100.pdf}
    };
    % \node[left = 0cm of T100.north west, anchor = north west, yshift = -0.15cm] (C) {
    %   \textbf{C}
    % };
    \node[above = -0.75cm of T100, anchor=south] () {
      \footnotesize{$T = 100$}
    };
    \node[right = -0.25cm of T100, anchor=south, rotate=270] () {
      \footnotesize{$M = 1000$}
    };

    % 100 importance samples

    % disp rates, no cutoff
    \node[below = -1cm of Tnone] (Tnone_m100) {
      \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone-m100.pdf}    
    };
    % \node[left = 0cm of Tnone_m10.north west, anchor = north west, yshift = -0.15cm] {
    %   \textbf{B}
    % };

    % disp rates, 1000 cutoff
    \node[below = -1cm of T1000] (T1000_m100) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000-m100.pdf}
    };
  
    % disp rates, 100 cutoff
    \node[below = -1cm of T100] (T100_m100) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100-m100.pdf}
    };
    \node[right = -0.25cm of T100_m100, anchor=south, rotate=270] () {
      \footnotesize{$M = 100$}
    };

    % 10 importance samples

    % disp rates, no cutoff
    \node[below = -1cm of Tnone_m100] (Tnone_m10) {
      \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone-m10.pdf}    
    };
    % \node[left = 0cm of Tnone_m1.north west, anchor = north west, yshift = -0.15cm] {
    %   \textbf{C}
    % };
    
    % disp rates, 1000 cutoff
    \node[below = -1cm of T1000_m100] (T1000_m10) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000-m10.pdf}
    };
  
    % disp rates, 100 cutoff
    \node[below = -1cm of T100_m100] (T100_m10) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100-m10.pdf}
    };
    \node[right = -0.25cm of T100_m10, anchor=south, rotate=270] () {
      \footnotesize{$M = 10$}
    };
 
    % 1 importance samples

    % disp rates, no cutoff
    \node[below = -1cm of Tnone_m10] (Tnone_m1) {
      \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone-m1.pdf}    
    };
    % \node[left = 0cm of Tnone_m1.north west, anchor = north west, yshift = -0.15cm] {
    %   \textbf{C}
    % };
    
    % disp rates, 1000 cutoff
    \node[below = -1cm of T1000_m10] (T1000_m1) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000-m1.pdf}
    };
  
    % disp rates, 100 cutoff
    \node[below = -1cm of T100_m10] (T100_m1) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100-m1.pdf}
    };
    \node[right = -0.25cm of T100_m1, anchor=south, rotate=270] () {
      \footnotesize{$M = 1$}
    };    

\end{tikzpicture}  

\end{center}

\caption{
\textbf{The effect of importance sampling on dispersal estimates.}
Importance sampling more trees, $M$, at each locus brings the dispersal estimate from the inferred trees closer to the dispersal estimate from the true trees and lowers the variance in dispersal estimates across replicates. See Figure \ref{fig:sims} for more information, where $M=1000$ (top row).
}

\refstepcounter{SIfig}\label{fig:sup_sigma_imp}
\end{figure}

% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% %simulation dispersal estimates -- effect of number of loci
% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% \begin{figure}[!hb]
%   %\captionsetup{labelformat=empty}
%   \begin{center}
%   \begin{tikzpicture}[remember picture]
  
%     % 100 loci

%     % disp rates, no cutoff
%     \node[] (Tnone) {
%       \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone.pdf}    
%     };
%     % \node[left = 0cm of Tnone.north west, anchor = north west, yshift = -0.15cm] (A) {
%     %   \textbf{A}
%     % };
%     \node[above = -0.75cm of Tnone, anchor=south] () {
%       \footnotesize{$T = \infty$}
%     };
  
%     % disp rates, 1000 cutoff
%     \node[right = -0.2cm of Tnone] (T1000) {
%       \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000.pdf}
%     };
%     % \node[left = 0cm of T1000.north west, anchor = north west, yshift = -0.15cm] (B) {
%     %   \textbf{B}
%     % };
%     \node[above = -0.75cm of T1000, anchor=south] () {
%       \footnotesize{$T = 1000$}
%     };
  
%     % disp rates, 100 cutoff
%     \node[right = -0.2cm of T1000] (T100) {
%       \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100.pdf}
%     };
%     % \node[left = 0cm of T100.north west, anchor = north west, yshift = -0.15cm] (C) {
%     %   \textbf{C}
%     % };
%     \node[above = -0.75cm of T100, anchor=south] () {
%       \footnotesize{$T = 100$}
%     };
%     \node[right = -0.25cm of T100, anchor=south, rotate=270] () {
%       \footnotesize{$L = 100$}
%     };

%     % 10 loci

%     % disp rates, no cutoff
%     \node[below = -1cm of Tnone] (Tnone_m10) {
%       \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone-10n.pdf}    
%     };
%     % \node[left = 0cm of Tnone_m10.north west, anchor = north west, yshift = -0.15cm] {
%     %   \textbf{B}
%     % };

%     % disp rates, 1000 cutoff
%     \node[below = -1cm of T1000] (T1000_m10) {
%       \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000-10n.pdf}
%     };
  
%     % disp rates, 100 cutoff
%     \node[below = -1cm of T100] (T100_m10) {
%       \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100-10n.pdf}
%     };
%     \node[right = -0.25cm of T100_m10, anchor=south, rotate=270] () {
%       \footnotesize{$L = 10$}
%     };

%     % 1 locus

%     % disp rates, no cutoff
%     \node[below = -1cm of Tnone_m10] (Tnone_m1) {
%       \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone-1n.pdf}    
%     };
%     % \node[left = 0cm of Tnone_m1.north west, anchor = north west, yshift = -0.15cm] {
%     %   \textbf{C}
%     % };
    
%     % disp rates, 1000 cutoff
%     \node[below = -1cm of T1000_m10] (T1000_m1) {
%       \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000-1n.pdf}
%     };
  
%     % disp rates, 100 cutoff
%     \node[below = -1cm of T100_m10] (T100_m1) {
%       \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100-1n.pdf}
%     };
%     \node[right = -0.25cm of T100_m1, anchor=south, rotate=270] () {
%       \footnotesize{$L = 1$}
%     };

% \end{tikzpicture}  

% \end{center}

% \caption{
% \textbf{The effect of the number of sampled loci on dispersal estimates.}
% Using more loci tends to reduce the variation in estimates across replicates at smaller cutoff times.
%  To speed up calculations, here we use $M=100$ importance samples. 
%  See Figure \ref{fig:sims} for more information, where $M=1000$ and we use every 100th locus. 
% \textcolor{red}{[would need to estimate dispersal rate from true trees with particular number of loci -- currently using every 100th locus]}
% }

% \label{fig:sup_sigma_loci}
% \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%simulation dispersal estimates -- effect of number of samples
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]
  %\captionsetup{labelformat=empty}
  \begin{center}
  \begin{tikzpicture}[remember picture]
  
    % 100 samples

    % disp rates, no cutoff
    \node[] (Tnone) {
      \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone-100k.pdf}    
    };
    % \node[left = 0cm of Tnone.north west, anchor = north west, yshift = -0.15cm] (A) {
    %   \textbf{A}
    % };
    \node[above = -0.75cm of Tnone, anchor=south] () {
      \footnotesize{$T = \infty$}
    };
  
    % disp rates, 1000 cutoff
    \node[right = -0.2cm of Tnone] (T1000) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000-100k.pdf}
    };
    % \node[left = 0cm of T1000.north west, anchor = north west, yshift = -0.15cm] (B) {
    %   \textbf{B}
    % };
    \node[above = -0.75cm of T1000, anchor=south] () {
      \footnotesize{$T = 1000$}
    };
  
    % disp rates, 100 cutoff
    \node[right = -0.2cm of T1000] (T100) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100-100k.pdf}
    };
    % \node[left = 0cm of T100.north west, anchor = north west, yshift = -0.15cm] (C) {
    %   \textbf{C}
    % };
    \node[above = -0.75cm of T100, anchor=south] () {
      \footnotesize{$T = 100$}
    };
    \node[right = -0.25cm of T100, anchor=south, rotate=270] () {
      \footnotesize{$k = 100$}
    };

    % 50 samples

    % disp rates, no cutoff
    \node[below = -1cm of Tnone] (Tnone_m10) {
      \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone-50k.pdf}    
    };
    % \node[left = 0cm of Tnone_m10.north west, anchor = north west, yshift = -0.15cm] {
    %   \textbf{B}
    % };

    % disp rates, 1000 cutoff
    \node[below = -1cm of T1000] (T1000_m10) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000-50k.pdf}
    };
  
    % disp rates, 100 cutoff
    \node[below = -1cm of T100] (T100_m10) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100-50k.pdf}
    };
    \node[right = -0.25cm of T100_m10, anchor=south, rotate=270] () {
      \footnotesize{$k = 50$}
    };

    % 25 samples

    % disp rates, no cutoff
    \node[below = -1cm of Tnone_m10] (Tnone_m1) {
      \includegraphics[height=0.39\linewidth, trim = {0cm 0 1cm 0}, clip]{disp-Tnone-25k.pdf}    
    };
    % \node[left = 0cm of Tnone_m1.north west, anchor = north west, yshift = -0.15cm] {
    %   \textbf{C}
    % };
    
    % disp rates, 1000 cutoff
    \node[below = -1cm of T1000_m10] (T1000_m1) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T1000-25k.pdf}
    };
  
    % disp rates, 100 cutoff
    \node[below = -1cm of T100_m10] (T100_m1) {
      \includegraphics[height=0.39\linewidth, trim = {1.5cm 0 1cm 0}, clip]{disp-T100-25k.pdf}
    };
    \node[right = -0.25cm of T100_m1, anchor=south, rotate=270] () {
      \footnotesize{$k = 25$}
    };

\end{tikzpicture}  

\end{center}

\caption{
\textbf{The effect of the number of samples on dispersal estimates.}
Using more samples tends to increase the dispersal estimate from both true and inferred trees and reduce the variation in estimates across replicates at smaller cutoff times.
See Figure \ref{fig:sims} for more information, where $k=50$, $M=1000$, and we sample every 100th locus. 
To isolate the effect of larger trees (rather than more trees), here we use the first $L=100$ sampled loci.
To speed up calculations, here we use $M=100$ importance samples at each locus. 
}

\refstepcounter{SIfig}\label{fig:sup_sigma_samples}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%blup locations
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]

  \begin{tikzpicture}
    \node[] (top-left) {
      \includegraphics[width=0.5\textwidth]{rmse-single-lineages.pdf}
    }; 
    \node[above = -0.5cm of top-left, anchor=south] () {
      \footnotesize{individual loci}
    };
    \node[right = -0.75cm of top-left] (top-right) {
      \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds.pdf}
    };
    \node[above = -0.5cm of top-right, anchor=south] () {
      \footnotesize{mean across loci}
    };
    \node[right = -0.75cm of top-right, anchor=south, rotate=270] () {
      \footnotesize{MLE}
    };
    \node[below = -0.5cm of top-left] () {
      \includegraphics[width=0.5\textwidth]{rmse-single-lineages-blup.pdf}
    };
    \node[below = -0.5cm of top-right] (bottom-right) {
      \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds-blup.pdf}
    };
    \node[right = -0.75cm of bottom-right, anchor=south, rotate=270] () {
      \footnotesize{BLUP}
    };
  \end{tikzpicture}

\caption{
\textbf{Best linear unbiased predictor (BLUP) of ancestor locations.} The top row shows error in the MLE ancestor locations (orange), as in Figure \ref{fig:sims}, while the bottom row shows the error in the BLUP ancestor locations (orange). The methods are very comparable, though the BLUP method may do slightly worse at estimating the mean ancestor location at deep times.
}

\refstepcounter{SIfig}\label{fig:blups}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%mle locations - importance samples
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]

  \begin{center}
  \begin{tikzpicture}[remember picture]
  
    % M = 1000
  
    \node[] (single-1000) {
      \includegraphics[width=0.5\textwidth]{rmse-single-lineages.pdf}
    };
  
    \node[above = -0.5cm of single-1000, anchor=south] () {
      \footnotesize{individual loci}
    };
  
    \node[right=-0.5cm of single-1000] (cloud-1000) {
      \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds.pdf}
    };
  
    \node[above = -0.5cm of cloud-1000, anchor=south] () {
      \footnotesize{mean across loci}
    };
  
    \node[right = -0.75cm of cloud-1000, anchor=south, rotate=270] () {
      \footnotesize{$M = 1000$}
    };
  
    % M = 100
  
    \node[below = -0.5cm of single-1000] (single-100) {
      \includegraphics[width=0.5\textwidth]{rmse-single-lineages-100m.pdf}
    };
  
    \node[right=-0.5cm of single-100] (cloud-100) {
      \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds-100m.pdf}
    };
  
    \node[right = -0.75cm of cloud-100, anchor=south, rotate=270] () {
      \footnotesize{$M = 100$}
    };
  
    % M = 10
  
    \node[below = -0.5cm of single-100] (single-10) {
      \includegraphics[width=0.5\textwidth]{rmse-single-lineages-10m.pdf}
    };
  
    \node[right=-0.5cm of single-10] (cloud-10) {
      \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds-10m.pdf}
    };
  
    \node[right = -0.75cm of cloud-10, anchor=south, rotate=270] () {
      \footnotesize{$M = 10$}
    };

    % M = 1
  
    \node[below = -0.5cm of single-10] (single-1) {
      \includegraphics[width=0.5\textwidth]{rmse-single-lineages-1m.pdf}
    };
  
    \node[right=-0.5cm of single-1] (cloud-1) {
      \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds-1m.pdf}
    };
  
    \node[right = -0.75cm of cloud-1, anchor=south, rotate=270] () {
      \footnotesize{$M = 1$}
    };

  \end{tikzpicture}
  \end{center}
  
  \caption{
  \textbf{The effect of importance samples on the accuracy of ancestor locations.} There is essentially no effect of importance sampling on the error in location estimates. See Figure \ref{fig:sims} for more information, where $M=1000$ (top row).
  }
  
  \refstepcounter{SIfig}\label{fig:anc-locs-M}
  \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%mle locations - sample size
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]

\begin{center}
\begin{tikzpicture}[remember picture]

  % k = 100

  \node[] (single-100) {
    \includegraphics[width=0.5\textwidth]{rmse-single-lineages-100k.pdf}
  };

  \node[above = -0.5cm of single-100, anchor=south] () {
    \footnotesize{individual loci}
  };

  \node[right=-0.5cm of single-100] (cloud-100) {
    \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds-100k.pdf}
  };

  \node[above = -0.5cm of cloud-100, anchor=south] () {
    \footnotesize{mean across loci}
  };

  \node[right = -0.75cm of cloud-100, anchor=south, rotate=270] () {
    \footnotesize{$k = 100$}
  };

  % k = 50

  \node[below = -0.5cm of single-100] (single-50) {
    \includegraphics[width=0.5\textwidth]{rmse-single-lineages-50k.pdf}
  };

  \node[right=-0.5cm of single-50] (cloud-50) {
    \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds-50k.pdf}
  };

  \node[right = -0.75cm of cloud-50, anchor=south, rotate=270] () {
    \footnotesize{$k = 50$}
  };

  % k = 25

  \node[below = -0.5cm of single-50] (single-25) {
    \includegraphics[width=0.5\textwidth]{rmse-single-lineages-25k.pdf}
  };

  \node[right=-0.5cm of single-25] (cloud-25) {
    \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds-25k.pdf}
  };

  \node[right = -0.75cm of cloud-25, anchor=south, rotate=270] () {
    \footnotesize{$k = 25$}
  };

\end{tikzpicture}
\end{center}

\caption{
\textbf{The effect of sample size on the accuracy of ancestor locations.} See Figure \ref{fig:sims} for more information, where $k=50$, $M=1000$, and we sample every 100th locus. To isolate the effect of larger trees (rather than more trees), here we use the first $L=100$ sampled loci.
To speed up calculations, here we use $M=100$ importance samples at each locus.
}

\refstepcounter{SIfig}\label{fig:anc-locs-k}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%mle locations - loci
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% \begin{figure}[!htb]

%   \begin{center}
%   \begin{tikzpicture}[remember picture]
  
%     % k = 50
  
%     \node[] (single-50) {
%       \includegraphics[width=0.5\textwidth]{rmse-single-lineages.pdf}
%     };
  
%     \node[above = -0.75cm of single-50, anchor=south] () {
%       \footnotesize{individual loci}
%     };
  
%     \node[right=-0.5cm of single-50] (cloud-50) {
%       \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds.pdf}
%     };
  
%     \node[above = -0.75cm of cloud-50, anchor=south] () {
%       \footnotesize{mean across loci}
%     };
  
%     \node[right = -0.75cm of cloud-50, anchor=south, rotate=270] () {
%       \footnotesize{$n = 100$}
%     };
  
%     % k = 25
  
%     \node[below = -0.5cm of single-50] (single-25) {
%       \includegraphics[width=0.5\textwidth]{rmse-single-lineages-25k.pdf}
%     };
  
%     \node[right=-0.5cm of single-25] (cloud-25) {
%       \includegraphics[width=0.5\textwidth]{rmse-ancestor-clouds-25k.pdf}
%     };
  
%     \node[right = -0.75cm of cloud-25, anchor=south, rotate=270] () {
%       \footnotesize{$n = 10$}
%     };
  
%   \end{tikzpicture}
%   \end{center}
  
%   \caption{
%   \textbf{The effect of number of loci on ancestor locations.} See Figure \ref{fig:sims} for more information, where we use every 100th locus. 
%   }
  
%   \label{fig:anc-locs-n}
%   \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%inferred demography
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{figure}[!htb]

\includegraphics[width=\textwidth]{effective-population-sizes.pdf}

\caption{
\textbf{\texttt{Relate}-inferred demography of \textit{Arabidopsis thaliana}.}
}

\refstepcounter{SIfig}\label{fig:nes}
\end{figure}

\end{document}
